diff --git a/drivers/pci/setup-bus.c b/drivers/pci/setup-bus.c
index 90a455573149..a55606aba813 100644
--- a/drivers/pci/setup-bus.c
+++ b/drivers/pci/setup-bus.c
@@ -1480,7 +1480,7 @@ static void pci_bus_dump_resources(struct pci_bus *bus)
 	}
 }
 
-static int pci_bus_get_depth(struct pci_bus *bus)
+static int __maybe_unused pci_bus_get_depth(struct pci_bus *bus)
 {
 	int depth = 0;
 	struct pci_bus *child_bus;
@@ -1519,7 +1519,7 @@ void __init pci_realloc_get_opt(char *str)
 	else if (!strncmp(str, "on", 2))
 		pci_realloc_enable = user_enabled;
 }
-static bool pci_realloc_enabled(enum enable_type enable)
+static bool __maybe_unused pci_realloc_enabled(enum enable_type enable)
 {
 	return enable >= user_enabled;
 }
@@ -1563,7 +1563,7 @@ static enum enable_type pci_realloc_detect(struct pci_bus *bus,
 	return enable_local;
 }
 #else
-static enum enable_type pci_realloc_detect(struct pci_bus *bus,
+static enum enable_type __maybe_unused pci_realloc_detect(struct pci_bus *bus,
 			 enum enable_type enable_local)
 {
 	return enable_local;
@@ -1589,17 +1589,6 @@ void pci_assign_unassigned_root_bus_resources(struct pci_bus *bus)
 	int pci_try_num = 1;
 	enum enable_type enable_local;
 
-	/* don't realloc if asked to do so */
-	enable_local = pci_realloc_detect(bus, pci_realloc_enable);
-	if (pci_realloc_enabled(enable_local)) {
-		int max_depth = pci_bus_get_depth(bus);
-
-		pci_try_num = max_depth + 1;
-		dev_printk(KERN_DEBUG, &bus->dev,
-			   "max bus depth: %d pci_try_num: %d\n",
-			   max_depth, pci_try_num);
-	}
-
 again:
 	/*
 	 * last try will use add_list, otherwise will try good to have as
