diff --git a/fs/fuse/dir.c b/fs/fuse/dir.c
index 2d8071355674..b33f436ca313 100644
--- a/fs/fuse/dir.c
+++ b/fs/fuse/dir.c
@@ -512,6 +512,9 @@ static int fuse_create_open(struct inode *dir, struct dentry *entry,
 	if (err)
 		goto out_free_ff;
 
+	if (req->private_lower_rw_file != NULL)
+		ff->rw_lower_file = req->private_lower_rw_file;
+
 	err = -EIO;
 	if (!S_ISREG(outentry.attr.mode) || invalid_nodeid(outentry.nodeid) ||
 	    fuse_invalid_attr(&outentry.attr))
