diff --git a/fs/fuse/inode.c b/fs/fuse/inode.c
index 0ac3d6fdd28b..2ec4ecef361e 100644
--- a/fs/fuse/inode.c
+++ b/fs/fuse/inode.c
@@ -905,6 +905,18 @@ static void process_init_reply(struct fuse_conn *fc, struct fuse_req *req)
 				fc->async_dio = 1;
 			if (arg->flags & FUSE_WRITEBACK_CACHE)
 				fc->writeback_cache = 1;
+			if (arg->flags & FUSE_SHORTCIRCUIT) {
+				fc->writeback_cache = 0;
+				fc->shortcircuit_io = 1;
+				pr_info("FUSE: SHORTCIRCUIT enabled [%s : %d]!\n",
+					current->comm, current->pid);
+			}
+			if (arg->flags & FUSE_RESERVE_SPACE) {
+				fc->reserved_space_mb = arg->reserved_space_mb;
+				pr_info("FUSE: RESERVE_SPACE enabled [%s : %d]! %u\n",
+						current->comm, current->pid,
+						arg->reserved_space_mb);
+			}
 			if (arg->time_gran && arg->time_gran <= 1000000000)
 				fc->sb->s_time_gran = arg->time_gran;
 		} else {
