diff --git a/include/linux/iio/buffer.h b/include/linux/iio/buffer.h
index 519392763393..7a168d19a381 100644
--- a/include/linux/iio/buffer.h
+++ b/include/linux/iio/buffer.h
@@ -209,6 +209,8 @@ ssize_t iio_buffer_show_enable(struct device *dev,
 					   iio_buffer_show_enable,	\
 					   iio_buffer_store_enable)
 
+int iio_sw_buffer_preenable(struct iio_dev *indio_dev);
+
 bool iio_validate_scan_mask_onehot(struct iio_dev *indio_dev,
 	const unsigned long *mask);
 
