diff --git a/sound/soc/samsung/dma.h b/sound/soc/samsung/dma.h
index 085ef30f5ca2..690547626502 100644
--- a/sound/soc/samsung/dma.h
+++ b/sound/soc/samsung/dma.h
@@ -12,19 +12,25 @@
 #ifndef _S3C_AUDIO_H
 #define _S3C_AUDIO_H
 
-#include <sound/dmaengine_pcm.h>
-
 struct s3c_dma_params {
-	void *slave;				/* Channel ID */
+	struct s3c2410_dma_client *client;	/* stream identifier */
+	int channel;				/* Channel ID */
+        void *slave;                            /* Channel ID */
 	dma_addr_t dma_addr;
 	int dma_size;			/* Size of the DMA transfer */
+#ifdef CONFIG_ARM64
+	unsigned long ch;
+#else
+	unsigned ch;
+#endif
+	struct samsung_dma_ops *ops;
+	struct device *sec_dma_dev;	/* stream identifier */
 	char *ch_name;
-	struct snd_dmaengine_dai_dma_data dma_data;
+	bool esa_dma;
+	bool compr_dma;
 };
 
-void samsung_asoc_init_dma_data(struct snd_soc_dai *dai,
-				struct s3c_dma_params *playback,
-				struct s3c_dma_params *capture);
-int samsung_asoc_dma_platform_register(struct device *dev);
+int asoc_dma_platform_register(struct device *dev);
+void asoc_dma_platform_unregister(struct device *dev);
 
 #endif
