diff --git a/sound/soc/samsung/i2s-regs.h b/sound/soc/samsung/i2s-regs.h
index 821a50231002..017cb0a57ca7 100644
--- a/sound/soc/samsung/i2s-regs.h
+++ b/sound/soc/samsung/i2s-regs.h
@@ -33,7 +33,7 @@
 #define I2SLVL3ADDR	0x3c
 #define I2SSTR1		0x40
 #define I2SVER		0x44
-#define I2SFIC2		0x48
+#define I2SFIC1		0x48
 #define I2STDM		0x4c
 
 #define CON_RSTCLR		(1 << 31)
@@ -88,31 +88,47 @@
 #define MOD_C1DD_LHALF		(1 << 18)
 #define MOD_DC2_EN		(1 << 17)
 #define MOD_DC1_EN		(1 << 16)
+
 #define MOD_BLC_16BIT		(0 << 13)
 #define MOD_BLC_8BIT		(1 << 13)
 #define MOD_BLC_24BIT		(2 << 13)
 #define MOD_BLC_MASK		(3 << 13)
 
-#define MOD_IMS_SYSMUX		(1 << 10)
+#define MOD_CDCLKCON		(1 << 12)
 #define MOD_SLAVE		(1 << 11)
-#define MOD_TXONLY		(0 << 8)
-#define MOD_RXONLY		(1 << 8)
-#define MOD_TXRX		(2 << 8)
-#define MOD_MASK		(3 << 8)
+#define MOD_RCLKSRC		(1 << 10)
+
+/*
+ * H/W Guide suggests that it is not allowed to change TXR mode while I2S is
+ * running. So I2S should use dedicated TXR mode as TX RX simultaneous mode.
+ * Since it is not possible to know beforehand whether the I2S operation would
+ * be only TX/RX or simultaneous TX RX, it is better to always set TXR mode.
+ */
+#define MOD_TXR_TXONLY		2
+#define MOD_TXR_RXONLY		2
+#define MOD_TXR_TXRX		2
+
+#define MOD_TXR_SHIFT		8
+#define MOD_TXR_MASK		3
+
+#define MOD_LRP			(1 << 7)
 #define MOD_LRP_SHIFT		7
 #define MOD_LR_LLOW		0
 #define MOD_LR_RLOW		1
+
 #define MOD_SDF_SHIFT		5
 #define MOD_SDF_IIS		0
 #define MOD_SDF_MSB		1
 #define MOD_SDF_LSB		2
 #define MOD_SDF_MASK		3
+
 #define MOD_RCLK_SHIFT		3
 #define MOD_RCLK_256FS		0
 #define MOD_RCLK_512FS		1
 #define MOD_RCLK_384FS		2
 #define MOD_RCLK_768FS		3
 #define MOD_RCLK_MASK		3
+
 #define MOD_BCLK_SHIFT		1
 #define MOD_BCLK_32FS		0
 #define MOD_BCLK_48FS		1
@@ -121,9 +137,29 @@
 #define MOD_BCLK_MASK		3
 #define MOD_8BIT		(1 << 0)
 
+#define I2S_STR_MOD_LRP_SHIFT		15
+#define I2S_STR_MOD_LRP			BIT(I2S_STR_MOD_LRP_SHIFT)
+#define I2S_STR_MOD_CDCLKCON		MOD_CDCLKCON
+#define I2S_STR_MOD_SLAVE		MOD_SLAVE
+#define I2S_STR_MOD_RCLKSRC		MOD_RCLKSRC
+#define I2S_STR_MOD_TXR_SHIFT		MOD_TXR_SHIFT
+#define I2S_STR_MOD_TXR_MASK		MOD_TXR_MASK
+#define I2S_STR_MOD_SDF_SHIFT		6
+#define I2S_STR_MOD_RCLK_SHIFT		MOD_RCLK_SHIFT
+#define I2S_STR_MOD_RCLK_MASK		0x7
+#define I2S_STR_MOD_BCLK_SHIFT		0
+#define I2S_STR_MOD_BCLK_MASK		0x7
+
+#define EXYNOS5420_MOD_LRP		(1 << 15)
+#define EXYNOS5420_MOD_CDCLKCON		MOD_CDCLKCON
+#define EXYNOS5420_MOD_SLAVE		MOD_SLAVE
+#define EXYNOS5420_MOD_RCLKSRC		MOD_RCLKSRC
 #define EXYNOS5420_MOD_LRP_SHIFT	15
+#define EXYNOS5420_MOD_TXR_SHIFT	MOD_TXR_SHIFT
+#define EXYNOS5420_MOD_TXR_MASK		MOD_TXR_MASK
 #define EXYNOS5420_MOD_SDF_SHIFT	6
 #define EXYNOS5420_MOD_RCLK_SHIFT	4
+#define EXYNOS5420_MOD_RCLK_MASK	MOD_RCLK_MASK
 #define EXYNOS5420_MOD_BCLK_SHIFT	0
 #define EXYNOS5420_MOD_BCLK_64FS	4
 #define EXYNOS5420_MOD_BCLK_96FS	5
@@ -132,7 +168,23 @@
 #define EXYNOS5420_MOD_BCLK_256FS	8
 #define EXYNOS5420_MOD_BCLK_MASK	0xf
 
-#define MOD_CDCLKCON		(1 << 12)
+#define EXYNOS5430_MOD_LRP		EXYNOS5420_MOD_LRP
+#define EXYNOS5430_MOD_CDCLKCON		(1 << 22)
+#define EXYNOS5430_MOD_SLAVE		(1 << 12)
+#define EXYNOS5430_MOD_RCLKSRC		(1 << 11)
+#define EXYNOS5430_MOD_LRP_SHIFT	EXYNOS5420_MOD_LRP_SHIFT
+#define EXYNOS5430_MOD_TXR_SHIFT	9
+#define EXYNOS5430_MOD_TXR_MASK		EXYNOS5420_MOD_TXR_MASK
+#define EXYNOS5430_MOD_SDF_SHIFT	7
+#define EXYNOS5430_MOD_RCLK_SHIFT	4
+#define EXYNOS5430_MOD_RCLK_MASK	7
+#define EXYNOS5430_MOD_RCLK_64FS	4
+#define EXYNOS5430_MOD_RCLK_128FS	5
+#define EXYNOS5430_MOD_RCLK_96FS	6
+#define EXYNOS5430_MOD_RCLK_192FS	7
+#define EXYNOS5430_MOD_BCLK_SHIFT	EXYNOS5420_MOD_BCLK_SHIFT
+#define EXYNOS5430_MOD_BCLK_MASK	EXYNOS5420_MOD_BCLK_MASK
+
 
 #define PSR_PSREN		(1 << 15)
 
@@ -157,6 +209,20 @@
 #define I2SSIZE_TRNMSK		(0xffff)
 #define I2SSIZE_SHIFT		(16)
 
+#define TDM_LRCLK_WIDTH_SHIFT	12
+#define TDM_LRCLK_WIDTH_MASK	0xFF
+#define TDM_RX_SLOTS_SHIFT	8
+#define TDM_RX_SLOTS_MASK	7
+#define TDM_TX_SLOTS_SHIFT	4
+#define TDM_TX_SLOTS_MASK	7
+#define TDM_MODE_MASK		(1 << 1)
+#define TDM_MODE_DSPA		(0 << 1)
+#define TDM_MODE_DSPB		(1 << 1)
+#define TDM_ENABLE		(1 << 0)
+
+#define TDM_RX_SLOTS_MAX	2
+#define TDM_TX_SLOTS_MAX	8
+
 #endif /* __SND_SOC_SAMSUNG_I2S_REGS_H */
 
 
