diff --git a/drivers/gpu/arm/t8xx/r7p0/backend/gpu/mali_kbase_cache_policy_backend.c b/drivers/gpu/arm/t8xx/r7p0/backend/gpu/mali_kbase_cache_policy_backend.c
new file mode 100644
index 000000000000..92a14fa1bae1
--- /dev/null
+++ b/drivers/gpu/arm/t8xx/r7p0/backend/gpu/mali_kbase_cache_policy_backend.c
@@ -0,0 +1,22 @@
+/*
+ *
+ * (C) COPYRIGHT 2015 ARM Limited. All rights reserved.
+ *
+ * This program is free software and is provided to you under the terms of the
+ * GNU General Public License version 2 as published by the Free Software
+ * Foundation, and any use by you of this program is subject to the terms
+ * of such GNU licence.
+ *
+ * A copy of the licence is included with the program, and can also be obtained
+ * from Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA.
+ *
+ */
+
+
+
+#include "backend/gpu/mali_kbase_cache_policy_backend.h"
+#include <backend/gpu/mali_kbase_pm_internal.h>
+#include <backend/gpu/mali_kbase_device_internal.h>
+
+
