diff --git a/net/bridge/br_vlan.c b/net/bridge/br_vlan.c
index f1e23e78fece..97b8ddf57363 100644
--- a/net/bridge/br_vlan.c
+++ b/net/bridge/br_vlan.c
@@ -360,12 +360,6 @@ void br_vlan_flush(struct net_bridge *br)
 	struct net_port_vlans *pv;
 
 	ASSERT_RTNL();
-
-	/* delete auto-added default pvid local fdb before flushing vlans
-	 * otherwise it will be leaked on bridge device init failure
-	 */
-	br_fdb_delete_by_port(br, NULL, 1);
-
 	pv = rtnl_dereference(br->vlan_info);
 	if (!pv)
 		return;
