diff --git a/.gitignore b/.gitignore
index 6ed8833731da..7a1121c1da3c 100644
--- a/.gitignore
+++ b/.gitignore
@@ -13,6 +13,7 @@
 *.o
 *.o.*
 *.a
+*.bin
 *.s
 *.ko
 *.so
@@ -23,7 +24,6 @@
 *.symtypes
 *.order
 *.elf
-*.bin
 *.gz
 *.bz2
 *.lzma
@@ -32,13 +32,13 @@
 *.lzo
 *.patch
 *.gcno
-modules.builtin
 Module.symvers
 *.dwo
 
 #
 # Top-level generic files
 #
+build_kernel.sh
 /tags
 /TAGS
 /linux
@@ -46,6 +46,7 @@ Module.symvers
 /vmlinuz
 /System.map
 /Module.markers
+/out
 
 #
 # Debian directory (make deb-pkg)
@@ -99,3 +100,9 @@ all.config
 
 # fetched Android config fragments
 android/configs/android-*.cfg
+
+# Gator Daemon files
+tools/gator/daemon/defaults_xml.h
+tools/gator/daemon/escape
+tools/gator/daemon/events.xml
+tools/gator/daemon/events_xml.h
