diff --git a/drivers/of/platform.c b/drivers/of/platform.c
index 3b64d0bf5bba..00a84bfbab3e 100644
--- a/drivers/of/platform.c
+++ b/drivers/of/platform.c
@@ -28,6 +28,9 @@ const struct of_device_id of_default_bus_match_table[] = {
 #ifdef CONFIG_ARM_AMBA
 	{ .compatible = "arm,amba-bus", },
 #endif /* CONFIG_ARM_AMBA */
+#ifdef CONFIG_EXYNOS_IOMMU
+	{ .compatible = "samsung,exynos-iommu-bus", },
+#endif
 	{} /* Empty terminated list */
 };
 
