diff --git a/include/dt-bindings/thermal/thermal.h b/include/dt-bindings/thermal/thermal.h
index 59822a995858..b6d8fec82b11 100644
--- a/include/dt-bindings/thermal/thermal.h
+++ b/include/dt-bindings/thermal/thermal.h
@@ -12,6 +12,15 @@
 
 /* On cooling devices upper and lower limits */
 #define THERMAL_NO_LIMIT		(-1UL)
+#define TABLE_END			(~1)
+
+#define THROTTLE_ACTIVE			(1)
+#define THROTTLE_PASSIVE		(2)
+#define SW_TRIP				(3)
+#define HW_TRIP				(4)
+
+#define DISABLE				(0)
+#define ENABLE				(1)
 
 #endif
 
