diff --git a/net/netfilter/interceptor/package_version.h b/net/netfilter/interceptor/package_version.h
new file mode 100644
index 000000000000..3fbcb6a412fb
--- /dev/null
+++ b/net/netfilter/interceptor/package_version.h
@@ -0,0 +1,13 @@
+/**
+   @copyright
+   Copyright (c) 2013 - 2015, INSIDE Secure Oy. All rights reserved.
+*/
+
+#ifndef PACKAGE_VERSION_H
+#define PACKAGE_VERSION_H
+
+#define PACKAGE_VERSION \
+        "939-23012-001-quicksec-vpnclient-for-android-3.5-4155-1a8773" \
+        "03f36b-4155 " \
+
+#endif
