diff --git a/net/netfilter/interceptor/public_defs.h b/net/netfilter/interceptor/public_defs.h
new file mode 100644
index 000000000000..ec85425abdc7
--- /dev/null
+++ b/net/netfilter/interceptor/public_defs.h
@@ -0,0 +1,15 @@
+/**
+   @copyright
+   Copyright (c) 2013, INSIDE Secure Oy. All rights reserved.
+*/
+
+#ifndef PUBLIC_DEFS_H
+#define PUBLIC_DEFS_H
+
+#ifdef __KERNEL__
+#include <linux/types.h>
+#else
+#include <stdint.h>
+#endif
+
+#endif /* PUBLIC_DEFS_H */
