diff --git a/arch/arm64/crypto/first_file_asm.c b/arch/arm64/crypto/first_file_asm.c
new file mode 100644
index 000000000000..37dacfb67537
--- /dev/null
+++ b/arch/arm64/crypto/first_file_asm.c
@@ -0,0 +1,20 @@
+#include <linux/init.h>
+
+const int first_crypto_asm_rodata = 10;
+int       first_crypto_asm_data   = 20;
+
+
+void first_crypto_asm_text (void) __attribute__((unused));
+void first_crypto_asm_text (void)
+{
+}
+
+void __init first_crypto_asm_init(void) __attribute__((unused));
+void __init first_crypto_asm_init(void)
+{
+}
+
+void __exit first_crypto_asm_exit(void) __attribute__((unused));
+void __exit first_crypto_asm_exit(void)
+{
+}
