diff --git a/drivers/media/platform/exynos/fimg2d_v5/fimg2d_cache.c b/drivers/media/platform/exynos/fimg2d_v5/fimg2d_cache.c
new file mode 100644
index 000000000000..4d37a1b53a35
--- /dev/null
+++ b/drivers/media/platform/exynos/fimg2d_v5/fimg2d_cache.c
@@ -0,0 +1,12 @@
+/* linux/drivers/media/video/exynos/fimg2d_v5/fimg2d_cache.c
+ *
+ * Copyright (c) 2015 Samsung Electronics Co., Ltd.
+ *	http://www.samsung.com/
+ *
+ * Samsung Graphics 2D driver
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+*/
+
