diff --git a/include/linux/backing-dev.h b/include/linux/backing-dev.h
index 7a59857c2700..12764cb9e2b3 100644
--- a/include/linux/backing-dev.h
+++ b/include/linux/backing-dev.h
@@ -89,6 +89,10 @@ struct backing_dev_info {
 
 	struct fprop_local_percpu completions;
 	int dirty_exceeded;
+	/* approximate write throttle statistics - updated at each throttling */
+	unsigned long last_thresh;  /* global/bdi thresh at the last throttle */
+	unsigned long last_nr_dirty; /* global/bdi dirty at the last throttle */
+	unsigned long paused_total; /* approximated sum of pauses. in jiffies */
 
 	unsigned int min_ratio;
 	unsigned int max_ratio, max_prop_frac;
