diff --git a/include/linux/device.h b/include/linux/device.h
index 452cd0cfea59..db9fd73e35f1 100644
--- a/include/linux/device.h
+++ b/include/linux/device.h
@@ -718,7 +718,6 @@ struct acpi_dev_node {
  * @iommu_group: IOMMU group the device belongs to.
  *
  * @offline_disabled: If set, the device is permanently online.
- * @offline:	Set after successful invocation of bus type's .offline().
  *
  * At the lowest level, every device in a Linux system is represented by an
  * instance of struct device. The device structure contains the information
@@ -796,7 +795,6 @@ struct device {
 	struct iommu_group	*iommu_group;
 
 	bool			offline_disabled:1;
-	bool			offline:1;
 };
 
 static inline struct device *kobj_to_dev(struct kobject *kobj)
@@ -912,6 +910,13 @@ static inline void device_unlock(struct device *dev)
 	mutex_unlock(&dev->mutex);
 }
 
+static inline struct device_node *dev_of_node(struct device *dev)
+{
+	if (!IS_ENABLED(CONFIG_OF))
+		return NULL;
+	return dev->of_node;
+}
+
 void driver_init(void);
 
 /*
