diff --git a/include/linux/dma-buf.h b/include/linux/dma-buf.h
index 694e1fe1c4b4..cb710072cd67 100644
--- a/include/linux/dma-buf.h
+++ b/include/linux/dma-buf.h
@@ -211,4 +211,5 @@ void *dma_buf_vmap(struct dma_buf *);
 void dma_buf_vunmap(struct dma_buf *, void *vaddr);
 int dma_buf_debugfs_create_file(const char *name,
 				int (*write)(struct seq_file *));
+struct dma_buf *get_dma_buf_file(struct file *);
 #endif /* __DMA_BUF_H__ */
