diff --git a/include/linux/if_pppox.h b/include/linux/if_pppox.h
index 2e9ba2f6014c..6932b77ac434 100644
--- a/include/linux/if_pppox.h
+++ b/include/linux/if_pppox.h
@@ -58,6 +58,7 @@ struct pppopns_opt {
 	__u32		xmit_sequence;
 	void		(*data_ready)(struct sock *sk_raw);
 	int		(*backlog_rcv)(struct sock *sk_raw, struct sk_buff *skb);
+	int 		ppp_flags;
 };
 
 #include <net/sock.h>
@@ -73,6 +74,8 @@ struct pppox_sock {
 		struct pppolac_opt lac;
 		struct pppopns_opt pns;
 	} proto;
+	struct timer_list recv_queue_timer;
+	spinlock_t recv_queue_lock;
 	__be16			num;
 };
 #define pppoe_dev	proto.pppoe.dev
