diff --git a/include/linux/jbd2.h b/include/linux/jbd2.h
index 7a9181af7367..c6d7e34ce007 100644
--- a/include/linux/jbd2.h
+++ b/include/linux/jbd2.h
@@ -1008,6 +1008,10 @@ struct journal_s
 						 * data write error in ordered
 						 * mode */
 #define JBD2_REC_ERR	0x080	/* The errno in the sb has been recorded */
+#ifdef CONFIG_JOURNAL_DATA_TAG
+#define JBD2_JOURNAL_TAG       0x800   /* Journaling is working in journal
+					* data tagging mode */
+#endif
 
 /*
  * Function declarations for the journaling transaction and buffer
