diff --git a/include/linux/kern_levels.h b/include/linux/kern_levels.h
index 866caaa9e2bb..862038d386be 100644
--- a/include/linux/kern_levels.h
+++ b/include/linux/kern_levels.h
@@ -15,6 +15,21 @@
 
 #define KERN_DEFAULT	KERN_SOH "d"	/* the default kernel loglevel */
 
+#ifdef CONFIG_SEC_DEBUG_AUTO_SUMMARY
+
+#define KERN_AUTO	KERN_SOH		/* auto-summary messages */
+#define KERN_AUTO1	"B"
+#define KERN_AUTO2	"C"
+#define KERN_AUTO3	"D"
+#define KERN_AUTO4	"E"
+#define KERN_AUTO5	"F"
+#define KERN_AUTO6	"G"
+#define KERN_AUTO7	"H"
+#define KERN_AUTO8	"I"
+#define KERN_AUTO9	"J"
+
+#endif
+
 /*
  * Annotation for a "continued" line of log printout (only done after a
  * line that had no enclosing \n). Only to be used by core/arch code
@@ -22,4 +37,8 @@
  */
 #define KERN_CONT	""
 
+#ifdef CONFIG_SEC_DEBUG_AUTO_SUMMARY
+#define LOGLEVEL_PR_AUTO_BASE	90	/* base log level for auto summary */
+#endif
+
 #endif
