diff --git a/include/linux/mmzone.h b/include/linux/mmzone.h
index 2594de31a1c7..00ed67d04cc3 100644
--- a/include/linux/mmzone.h
+++ b/include/linux/mmzone.h
@@ -64,9 +64,11 @@ enum {
 };
 
 #ifdef CONFIG_CMA
+bool is_cma_pageblock(struct page *page);
 #  define is_migrate_cma(migratetype) unlikely((migratetype) == MIGRATE_CMA)
 #  define is_migrate_cma_page(_page) (get_pageblock_migratetype(_page) == MIGRATE_CMA)
 #else
+#  define is_cma_pageblock(page) false
 #  define is_migrate_cma(migratetype) false
 #  define is_migrate_cma_page(_page) false
 #endif
@@ -361,6 +363,9 @@ struct zone {
 	 */
 	unsigned long		dirty_balance_reserve;
 
+#ifdef CONFIG_CMA
+	bool			cma_alloc;
+#endif
 #ifndef CONFIG_SPARSEMEM
 	/*
 	 * Flags for a pageblock_nr_pages block. See pageblock-flags.h.
