diff --git a/include/linux/netdevice.h b/include/linux/netdevice.h
index 74a3ce168a3d..d1b4eee3a6ba 100644
--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -1714,6 +1714,10 @@ struct net_device {
 	struct lock_class_key *qdisc_tx_busylock;
 	int group;
 	struct pm_qos_request	pm_qos_req;
+
+#ifdef CONFIG_NETPM
+	bool netpm_use;
+#endif
 };
 #define to_net_dev(d) container_of(d, struct net_device, dev)
 
