diff --git a/include/linux/notifier.h b/include/linux/notifier.h
index d14a4c362465..498fb2f81f77 100644
--- a/include/linux/notifier.h
+++ b/include/linux/notifier.h
@@ -211,5 +211,9 @@ static inline int notifier_to_errno(int ret)
 
 extern struct blocking_notifier_head reboot_notifier_list;
 
+extern int show_mem_extra_notifier_register(struct notifier_block *nb);
+extern int show_mem_extra_notifier_unregister(struct notifier_block *nb);
+extern void show_mem_extra_call_notifiers(void);
+
 #endif /* __KERNEL__ */
 #endif /* _LINUX_NOTIFIER_H */
