diff --git a/include/linux/of_dma.h b/include/linux/of_dma.h
index 56bc026c143f..1cc81e1cfeba 100644
--- a/include/linux/of_dma.h
+++ b/include/linux/of_dma.h
@@ -43,7 +43,18 @@ extern struct dma_chan *of_dma_simple_xlate(struct of_phandle_args *dma_spec,
 		struct of_dma *ofdma);
 extern struct dma_chan *of_dma_xlate_by_chan_id(struct of_phandle_args *dma_spec,
 		struct of_dma *ofdma);
+extern unsigned int of_dma_get_mcode_addr(struct device_node *np);
+extern bool of_dma_secure_mode(struct device_node *np);
+extern void __iomem *of_dma_get_arwrapper_address(struct device_node *np, unsigned int num);
+extern void __iomem *of_dma_get_awwrapper_address(struct device_node *np, unsigned int num);
+extern void __iomem *of_dma_get_instwrapper_address(struct device_node *np);
+extern bool of_dma_get_wrapper_available(struct device_node *np);
+extern u64 of_dma_get_mask(struct device_node *np, char *name);
 #else
+static inline unsigned int of_dma_get_mcode_addr(struct device_node *np)
+{
+	return -ENODEV;
+}
 static inline int of_dma_controller_register(struct device_node *np,
 		struct dma_chan *(*of_dma_xlate)
 		(struct of_phandle_args *, struct of_dma *),
@@ -68,6 +79,35 @@ static inline struct dma_chan *of_dma_simple_xlate(struct of_phandle_args *dma_s
 	return NULL;
 }
 
+static inline void __iomem *of_dma_get_wrapper_address(struct device_node *np, unsigned int num)
+{
+	return NULL;
+}
+
+static inline void __iomem *of_dma_get_arwrapper_address(struct device_node *np, unsigned int num)
+{
+	return NULL;
+}
+
+static inline void __iomem *of_dma_get_awwrapper_address(struct device_node *np, unsigned int num)
+{
+	return NULL;
+}
+
+static inline void __iomem *of_dma_get_instwrapper_address(struct device_node *np)
+{
+	return NULL;
+}
+
+static inline bool of_dma_get_wrapper_available(struct device_node *np)
+{
+	return NULL;
+}
+
+static u64 of_dma_get_mask(struct device_node *np, char *name)
+{
+	return NULL;
+}
 #define of_dma_xlate_by_chan_id NULL
 
 #endif
