diff --git a/include/linux/pm_wakeup.h b/include/linux/pm_wakeup.h
index a0f70808d7f4..503a5ac00700 100644
--- a/include/linux/pm_wakeup.h
+++ b/include/linux/pm_wakeup.h
@@ -54,6 +54,10 @@ struct wakeup_source {
 	ktime_t last_time;
 	ktime_t start_prevent_time;
 	ktime_t prevent_sleep_time;
+#ifdef CONFIG_SEC_PM_DEBUG
+	ktime_t start_screen_off;
+	ktime_t time_while_screen_off;
+#endif
 	unsigned long		event_count;
 	unsigned long		active_count;
 	unsigned long		relax_count;
@@ -61,6 +65,9 @@ struct wakeup_source {
 	unsigned long		wakeup_count;
 	bool			active:1;
 	bool			autosleep_enabled:1;
+#ifdef CONFIG_SEC_PM_DEBUG
+	bool			is_screen_off:1;
+#endif
 };
 
 #ifdef CONFIG_PM_SLEEP
