diff --git a/include/linux/tcp.h b/include/linux/tcp.h
index 0c53c8efc8ab..e6be1eb9dbaf 100644
--- a/include/linux/tcp.h
+++ b/include/linux/tcp.h
@@ -283,6 +283,19 @@ struct tcp_sock {
 
 	int			linger2;
 
+/* Network Pacemaker */
+#ifdef CONFIG_NETPM
+	u8 netpm_netif;
+	u8 netpm_rbuf_flag;
+	u32 netpm_rtt_min;
+	u32 netpm_srtt;
+	u32 netpm_rttvar;
+	int netpm_cwnd_est;
+	int netpm_tcp_rmem_max;
+	int netpm_max_tput;
+	int netpm_rmem_max_curbdp;
+#endif
+
 /* Receiver side RTT estimation */
 	struct {
 		u32	rtt;
