diff --git a/include/linux/tty.h b/include/linux/tty.h
index 2ea62e9bb236..a450bdd605eb 100644
--- a/include/linux/tty.h
+++ b/include/linux/tty.h
@@ -258,6 +258,10 @@ struct tty_struct {
 	struct termiox *termiox;	/* May be NULL for unsupported */
 	char name[64];
 	struct pid *pgrp;		/* Protected by ctrl lock */
+	/*
+	 * Writes protected by both ctrl lock and legacy mutex, readers must use
+	 * at least one of them.
+	 */
 	struct pid *session;
 	unsigned long flags;
 	int count;
