diff --git a/include/linux/vm_event_item.h b/include/linux/vm_event_item.h
index 730334cdf037..d0ee6188e904 100644
--- a/include/linux/vm_event_item.h
+++ b/include/linux/vm_event_item.h
@@ -52,6 +52,10 @@ enum vm_event_item { PGPGIN, PGPGOUT, PSWPIN, PSWPOUT,
 		COMPACTMIGRATE_SCANNED, COMPACTFREE_SCANNED,
 		COMPACTISOLATED,
 		COMPACTSTALL, COMPACTFAIL, COMPACTSUCCESS,
+#ifdef CONFIG_SEC_PHCOMP		
+		COMPACTDEFERED, COMPACTCALLDEFER,
+		PHCOMPDEFERED, PHCOMPCALLDEFER,
+#endif		
 #endif
 #ifdef CONFIG_HUGETLB_PAGE
 		HTLB_BUDDY_PGALLOC, HTLB_BUDDY_PGALLOC_FAIL,
