diff --git a/include/linux/vmw_vmci_defs.h b/include/linux/vmw_vmci_defs.h
index 7023432013e8..65ac54c61c18 100644
--- a/include/linux/vmw_vmci_defs.h
+++ b/include/linux/vmw_vmci_defs.h
@@ -75,18 +75,9 @@ enum {
 
 /*
  * A single VMCI device has an upper limit of 128MB on the amount of
- * memory that can be used for queue pairs. Since each queue pair
- * consists of at least two pages, the memory limit also dictates the
- * number of queue pairs a guest can create.
+ * memory that can be used for queue pairs.
  */
 #define VMCI_MAX_GUEST_QP_MEMORY (128 * 1024 * 1024)
-#define VMCI_MAX_GUEST_QP_COUNT  (VMCI_MAX_GUEST_QP_MEMORY / PAGE_SIZE / 2)
-
-/*
- * There can be at most PAGE_SIZE doorbells since there is one doorbell
- * per byte in the doorbell bitmap page.
- */
-#define VMCI_MAX_GUEST_DOORBELL_COUNT PAGE_SIZE
 
 /*
  * Queues with pre-mapped data pages must be small, so that we don't pin
