diff --git a/include/linux/wakeup_reason.h b/include/linux/wakeup_reason.h
index d84d8c301546..a4fad4d59f19 100644
--- a/include/linux/wakeup_reason.h
+++ b/include/linux/wakeup_reason.h
@@ -23,6 +23,9 @@
 void log_wakeup_reason(int irq);
 int check_wakeup_reason(int irq);
 
+#ifdef CONFIG_ARCH_EXYNOS
+void update_wakeup_reason_stats(int irq, int eint);
+#endif
 #ifdef CONFIG_SUSPEND
 void log_suspend_abort_reason(const char *fmt, ...);
 #else
