diff --git a/block/ioctl.c b/block/ioctl.c
index 6c7bf903742f..9217dc058bb5 100644
--- a/block/ioctl.c
+++ b/block/ioctl.c
@@ -182,6 +182,8 @@ static int blk_ioctl_discard(struct block_device *bdev, uint64_t start,
 		return -EINVAL;
 	if (secure)
 		flags |= BLKDEV_DISCARD_SECURE;
+	printk("%s %d:%d %llu %llu", secure?"SECDIS":"DIS", MAJOR(bdev->bd_dev), MINOR(bdev->bd_dev), 
+		(unsigned long long)start, (unsigned long long)len);
 	return blkdev_issue_discard(bdev, start, len, GFP_KERNEL, flags);
 }
 
