diff --git a/block/scsi_ioctl.c b/block/scsi_ioctl.c
index 4a2735dce513..c916307e367b 100644
--- a/block/scsi_ioctl.c
+++ b/block/scsi_ioctl.c
@@ -719,6 +719,8 @@ int scsi_verify_blk_ioctl(struct block_device *bd, unsigned int cmd)
 	case SG_GET_RESERVED_SIZE:
 	case SG_SET_RESERVED_SIZE:
 	case SG_EMULATED_HOST:
+	case SCSI_IOCTL_SECURITY_PROTOCOL_IN:
+	case SCSI_IOCTL_SECURITY_PROTOCOL_OUT:
 		return 0;
 	case CDROM_GET_CAPABILITY:
 		/* Keep this until we remove the printk below.  udev sends it
