diff --git a/kernel/trace/blktrace.c b/kernel/trace/blktrace.c
index 0661dd3a4e8a..bd3e59ea8ea9 100644
--- a/kernel/trace/blktrace.c
+++ b/kernel/trace/blktrace.c
@@ -1918,6 +1918,8 @@ void blk_dump_cmd(char *buf, struct request *rq)
 	}
 }
 
+SIO_PATCH_VERSION(ftrace_discard_bugfix, 1, 0, "");
+
 void blk_fill_rwbs(char *rwbs, u32 rw, int bytes)
 {
 	int i = 0;
@@ -1925,10 +1927,10 @@ void blk_fill_rwbs(char *rwbs, u32 rw, int bytes)
 	if (rw & REQ_FLUSH)
 		rwbs[i++] = 'F';
 
-	if (rw & WRITE)
-		rwbs[i++] = 'W';
-	else if (rw & REQ_DISCARD)
+	if (rw & REQ_DISCARD)
 		rwbs[i++] = 'D';
+	else if (rw & WRITE)
+		rwbs[i++] = 'W';
 	else if (bytes)
 		rwbs[i++] = 'R';
 	else
