diff --git a/drivers/hid/hid-prodikeys.c b/drivers/hid/hid-prodikeys.c
index 7905d9f015ed..91fab975063c 100644
--- a/drivers/hid/hid-prodikeys.c
+++ b/drivers/hid/hid-prodikeys.c
@@ -557,14 +557,10 @@ static void pcmidi_setup_extra_keys(
 
 static int pcmidi_set_operational(struct pcmidi_snd *pm)
 {
-	int rc;
-
 	if (pm->ifnum != 1)
 		return 0; /* only set up ONCE for interace 1 */
 
-	rc = pcmidi_get_output_report(pm);
-	if (rc < 0)
-		return rc;
+	pcmidi_get_output_report(pm);
 	pcmidi_submit_output_report(pm, 0xc1);
 	return 0;
 }
@@ -693,11 +689,7 @@ static int pcmidi_snd_initialise(struct pcmidi_snd *pm)
 	spin_lock_init(&pm->rawmidi_in_lock);
 
 	init_sustain_timers(pm);
-	err = pcmidi_set_operational(pm);
-	if (err < 0) {
-		pk_error("failed to find output report\n");
-		goto fail_register;
-	}
+	pcmidi_set_operational(pm);
 
 	/* register it */
 	err = snd_card_register(card);
