diff --git a/drivers/usb/gadget/udc/lpc32xx_udc.c b/drivers/usb/gadget/udc/lpc32xx_udc.c
index 19daa3cf240b..5ce885068085 100644
--- a/drivers/usb/gadget/udc/lpc32xx_udc.c
+++ b/drivers/usb/gadget/udc/lpc32xx_udc.c
@@ -1227,11 +1227,11 @@ static void udc_pop_fifo(struct lpc32xx_udc *udc, u8 *data, u32 bytes)
 			tmp = readl(USBD_RXDATA(udc->udp_baseaddr));
 
 			bl = bytes - n;
-			if (bl > 4)
-				bl = 4;
+			if (bl > 3)
+				bl = 3;
 
 			for (i = 0; i < bl; i++)
-				data[n + i] = (u8) ((tmp >> (i * 8)) & 0xFF);
+				data[n + i] = (u8) ((tmp >> (n * 8)) & 0xFF);
 		}
 		break;
 
