diff --git a/drivers/input/mouse/trackpoint.h b/drivers/input/mouse/trackpoint.h
index d0e0136dcf0c..2d7be0435957 100644
--- a/drivers/input/mouse/trackpoint.h
+++ b/drivers/input/mouse/trackpoint.h
@@ -148,8 +148,7 @@ struct trackpoint_data
 #ifdef CONFIG_MOUSE_PS2_TRACKPOINT
 int trackpoint_detect(struct psmouse *psmouse, bool set_properties);
 #else
-static inline int trackpoint_detect(struct psmouse *psmouse,
-				    bool set_properties)
+inline int trackpoint_detect(struct psmouse *psmouse, bool set_properties)
 {
 	return -ENOSYS;
 }
