diff --git a/arch/arm64/mm/flush.c b/arch/arm64/mm/flush.c
index a90615baa529..c26b804015e8 100644
--- a/arch/arm64/mm/flush.c
+++ b/arch/arm64/mm/flush.c
@@ -70,6 +70,10 @@ void __sync_icache_dcache(pte_t pte, unsigned long addr)
 {
 	struct page *page = pte_page(pte);
 
+	/* no flushing needed for anonymous pages */
+	if (!page_mapping(page))
+		return;
+
 	if (!test_and_set_bit(PG_dcache_clean, &page->flags)) {
 		__flush_dcache_area(page_address(page),
 				PAGE_SIZE << compound_order(page));
