diff --git a/arch/arm/mm/mmu.c b/arch/arm/mm/mmu.c
index ff61335cbade..f2f56b7b48b9 100644
--- a/arch/arm/mm/mmu.c
+++ b/arch/arm/mm/mmu.c
@@ -1546,6 +1546,7 @@ void __init paging_init(const struct machine_desc *mdesc)
 {
 	void *zero_page;
 
+	set_memsize_kernel_type(MEMSIZE_KERNEL_PAGING);
 	build_mem_type_table();
 	prepare_page_table();
 	map_lowmem();
@@ -1563,4 +1564,5 @@ void __init paging_init(const struct machine_desc *mdesc)
 
 	empty_zero_page = virt_to_page(zero_page);
 	__flush_dcache_page(NULL, empty_zero_page);
+	set_memsize_kernel_type(MEMSIZE_KERNEL_OTHERS);
 }
