diff --git a/fs/jbd2/commit.c b/fs/jbd2/commit.c
index c531406ce94b..b3c311088e9e 100644
--- a/fs/jbd2/commit.c
+++ b/fs/jbd2/commit.c
@@ -156,6 +156,11 @@ static int journal_submit_commit_record(journal_t *journal,
 	set_buffer_uptodate(bh);
 	bh->b_end_io = journal_end_buffer_io_sync;
 
+#ifdef CONFIG_JOURNAL_DATA_TAG
+	if (journal->j_flags & JBD2_JOURNAL_TAG)
+		set_buffer_journal(bh);
+#endif
+
 	if (journal->j_flags & JBD2_BARRIER &&
 	    !JBD2_HAS_INCOMPAT_FEATURE(journal,
 				       JBD2_FEATURE_INCOMPAT_ASYNC_COMMIT))
@@ -740,6 +745,10 @@ start_journal_io:
 				clear_buffer_dirty(bh);
 				set_buffer_uptodate(bh);
 				bh->b_end_io = journal_end_buffer_io_sync;
+#ifdef CONFIG_JOURNAL_DATA_TAG
+				if (journal->j_flags & JBD2_JOURNAL_TAG)
+					set_buffer_journal(bh);
+#endif
 				submit_bh(WRITE_SYNC, bh);
 			}
 			cond_resched();
