diff --git a/fs/ext4/dir.c b/fs/ext4/dir.c
index 7a5428960751..40f74c154879 100644
--- a/fs/ext4/dir.c
+++ b/fs/ext4/dir.c
@@ -89,6 +89,10 @@ int __ext4_check_dir_entry(const char *function, unsigned int line,
 	else
 		return 0;
 
+	/* for debugging, sangwoo2.lee */
+	print_bh(dir->i_sb, bh, 0, EXT4_BLOCK_SIZE(dir->i_sb));
+	/* for debugging */
+
 	if (filp)
 		ext4_error_file(filp, function, line, bh->b_blocknr,
 				"bad entry in directory: %s - offset=%u, "
