diff --git a/drivers/rtc/rtc-sysfs.c b/drivers/rtc/rtc-sysfs.c
index babd43bf3ddc..d54867b8e8cf 100644
--- a/drivers/rtc/rtc-sysfs.c
+++ b/drivers/rtc/rtc-sysfs.c
@@ -122,6 +122,26 @@ hctosys_show(struct device *dev, struct device_attribute *attr, char *buf)
 }
 static DEVICE_ATTR_RO(hctosys);
 
+#if defined(CONFIG_RTC_ALARM_BOOT)
+static ssize_t
+alarm_boot_show(struct device *dev, struct device_attribute *attr,
+		char *buf)
+{
+	ssize_t retval;
+	struct rtc_wkalrm alm;
+
+	retval = rtc_get_alarm_boot(to_rtc_device(dev), &alm);
+	if (retval) {
+		retval = sprintf(buf, "%d", alm.enabled);
+		printk(" rtc_sysfs_show_wakealarm -- enabled? : %d\n", alm.enabled);
+		return retval;
+	}
+
+	return retval;
+}
+static DEVICE_ATTR_RO(alarm_boot);
+#endif
+
 static struct attribute *rtc_attrs[] = {
 	&dev_attr_name.attr,
 	&dev_attr_date.attr,
@@ -129,6 +149,9 @@ static struct attribute *rtc_attrs[] = {
 	&dev_attr_since_epoch.attr,
 	&dev_attr_max_user_freq.attr,
 	&dev_attr_hctosys.attr,
+#if defined(CONFIG_RTC_ALARM_BOOT)
+	&dev_attr_alarm_boot.attr,
+#endif
 	NULL,
 };
 ATTRIBUTE_GROUPS(rtc);
