diff --git a/kernel/futex.c b/kernel/futex.c
index cd9aec0b0fad..aa0da344608b 100644
--- a/kernel/futex.c
+++ b/kernel/futex.c
@@ -561,8 +561,8 @@ again:
 		 * applies. If this is really a shmem page then the page lock
 		 * will prevent unexpected transitions.
 		 */
-		lock_page(page_head);
-		shmem_swizzled = PageSwapCache(page_head) || page_head->mapping;
+		lock_page(page);
+		shmem_swizzled = PageSwapCache(page) || page->mapping;
 		unlock_page(page_head);
 		put_page(page_head);
 
@@ -2891,7 +2891,7 @@ SYSCALL_DEFINE3(get_robust_list, int, pid,
 	}
 
 	ret = -EPERM;
-	if (!ptrace_may_access(p, PTRACE_MODE_READ_REALCREDS))
+	if (!ptrace_may_access(p, PTRACE_MODE_READ))
 		goto err_unlock;
 
 	head = p->robust_list;
