diff --git a/kernel/notifier.c b/kernel/notifier.c
index da9cd288a0d2..be186d1cd899 100644
--- a/kernel/notifier.c
+++ b/kernel/notifier.c
@@ -5,6 +5,8 @@
 #include <linux/rcupdate.h>
 #include <linux/vmalloc.h>
 #include <linux/reboot.h>
+#include <linux/suspend.h>
+#include <linux/exynos-ss.h>
 
 /*
  *	Notifier list for kernel code which wants to be called
@@ -90,7 +92,11 @@ static int notifier_call_chain(struct notifier_block **nl,
 			continue;
 		}
 #endif
+		if (val == PM_SUSPEND_PREPARE || val == PM_POST_SUSPEND)
+			exynos_ss_suspend(nb->notifier_call, NULL, ESS_FLAG_IN);
 		ret = nb->notifier_call(nb, val, v);
+		if (val == PM_SUSPEND_PREPARE || val == PM_POST_SUSPEND)
+			exynos_ss_suspend(nb->notifier_call, NULL, ESS_FLAG_OUT);
 
 		if (nr_calls)
 			(*nr_calls)++;
