diff --git a/kernel/pid.c b/kernel/pid.c
index 50ad2e5b2963..e4b3013e2c53 100644
--- a/kernel/pid.c
+++ b/kernel/pid.c
@@ -528,7 +528,7 @@ pid_t __task_pid_nr_ns(struct task_struct *task, enum pid_type type,
 		if (type != PIDTYPE_PID) {
 			if (type == __PIDTYPE_TGID)
 				type = PIDTYPE_PID;
-			task = task->group_leader;
+ 			task = task->group_leader;
 		}
 		nr = pid_nr_ns(task->pids[type].pid, ns);
 	}
@@ -572,10 +572,12 @@ void __init pidhash_init(void)
 {
 	unsigned int i, pidhash_size;
 
+	set_memsize_kernel_type(MEMSIZE_KERNEL_PIDHASH);
 	pid_hash = alloc_large_system_hash("PID", sizeof(*pid_hash), 0, 18,
 					   HASH_EARLY | HASH_SMALL,
 					   &pidhash_shift, NULL,
 					   0, 4096);
+	set_memsize_kernel_type(MEMSIZE_KERNEL_OTHERS);
 	pidhash_size = 1U << pidhash_shift;
 
 	for (i = 0; i < pidhash_size; i++)
