diff --git a/kernel/resource.c b/kernel/resource.c
index 83658ca0e69c..0a0c47faedb8 100644
--- a/kernel/resource.c
+++ b/kernel/resource.c
@@ -171,7 +171,7 @@ static const struct file_operations proc_iomem_operations = {
 static int __init ioresources_init(void)
 {
 	proc_create("ioports", 0, NULL, &proc_ioports_operations);
-	proc_create("iomem", 0, NULL, &proc_iomem_operations);
+	proc_create("iomem", S_IRUSR, NULL, &proc_iomem_operations);
 	return 0;
 }
 __initcall(ioresources_init);
