diff --git a/drivers/staging/android/ashmem.c b/drivers/staging/android/ashmem.c
index 70c7f1648b4d..2fc4656f6692 100644
--- a/drivers/staging/android/ashmem.c
+++ b/drivers/staging/android/ashmem.c
@@ -365,8 +365,8 @@ static int ashmem_vmfile_mmap(struct file *file, struct vm_area_struct *vma)
 
 static unsigned long
 ashmem_vmfile_get_unmapped_area(struct file *file, unsigned long addr,
-				unsigned long len, unsigned long pgoff,
-				unsigned long flags)
+			unsigned long len, unsigned long pgoff,
+			unsigned long flags)
 {
 	return current->mm->get_unmapped_area(file, addr, len, pgoff, flags);
 }
@@ -424,7 +424,7 @@ static int ashmem_mmap(struct file *file, struct vm_area_struct *vma)
 			vmfile_fops = *vmfile->f_op;
 			vmfile_fops.mmap = ashmem_vmfile_mmap;
 			vmfile_fops.get_unmapped_area =
-					ashmem_vmfile_get_unmapped_area;
+				ashmem_vmfile_get_unmapped_area;
 		}
 		vmfile->f_op = &vmfile_fops;
 	}
