diff --git a/include/linux/mfd/samsung/s5m8767.h b/include/linux/mfd/samsung/s5m8767.h
index 243b58fec33d..306a95fc558c 100644
--- a/include/linux/mfd/samsung/s5m8767.h
+++ b/include/linux/mfd/samsung/s5m8767.h
@@ -183,29 +183,6 @@ enum s5m8767_regulators {
 	S5M8767_REG_MAX,
 };
 
-/* LDO_EN/BUCK_EN field in registers */
-#define S5M8767_ENCTRL_SHIFT		6
-#define S5M8767_ENCTRL_MASK		(0x3 << S5M8767_ENCTRL_SHIFT)
-
-/*
- * LDO_EN/BUCK_EN register value for controlling this Buck or LDO
- * by GPIO (PWREN, BUCKEN).
- */
-#define S5M8767_ENCTRL_USE_GPIO		0x1
-
-/*
- * Values for BUCK_RAMP field in DVS_RAMP register, matching raw values
- * in mV/us.
- */
-enum s5m8767_dvs_buck_ramp_values {
-	S5M8767_DVS_BUCK_RAMP_5		= 0x4,
-	S5M8767_DVS_BUCK_RAMP_10	= 0x9,
-	S5M8767_DVS_BUCK_RAMP_12_5	= 0xb,
-	S5M8767_DVS_BUCK_RAMP_25	= 0xd,
-	S5M8767_DVS_BUCK_RAMP_50	= 0xe,
-	S5M8767_DVS_BUCK_RAMP_100	= 0xf,
-};
-#define S5M8767_DVS_BUCK_RAMP_SHIFT	4
-#define S5M8767_DVS_BUCK_RAMP_MASK	(0xf << S5M8767_DVS_BUCK_RAMP_SHIFT)
+#define S5M8767_ENCTRL_SHIFT  6
 
 #endif /* __LINUX_MFD_S5M8767_H */
