diff --git a/drivers/net/usb/usbnet.c b/drivers/net/usb/usbnet.c
index fe4d84af5d59..ddda8e134db7 100644
--- a/drivers/net/usb/usbnet.c
+++ b/drivers/net/usb/usbnet.c
@@ -114,11 +114,6 @@ int usbnet_get_endpoints(struct usbnet *dev, struct usb_interface *intf)
 			int				intr = 0;
 
 			e = alt->endpoint + ep;
-
-			/* ignore endpoints which cannot transfer data */
-			if (!usb_endpoint_maxp(&e->desc))
-				continue;
-
 			switch (e->desc.bmAttributes) {
 			case USB_ENDPOINT_XFER_INT:
 				if (!usb_endpoint_dir_in(&e->desc))
@@ -352,8 +347,6 @@ void usbnet_update_max_qlen(struct usbnet *dev)
 {
 	enum usb_device_speed speed = dev->udev->speed;
 
-	if (!dev->rx_urb_size || !dev->hard_mtu)
-		goto insanity;
 	switch (speed) {
 	case USB_SPEED_HIGH:
 		dev->rx_qlen = MAX_QUEUE_MEMORY / dev->rx_urb_size;
@@ -369,7 +362,6 @@ void usbnet_update_max_qlen(struct usbnet *dev)
 		dev->tx_qlen = 5 * MAX_QUEUE_MEMORY / dev->hard_mtu;
 		break;
 	default:
-insanity:
 		dev->rx_qlen = dev->tx_qlen = 4;
 	}
 }
