diff --git a/include/linux/phy/phy.h b/include/linux/phy/phy.h
index 8cb6f815475b..c06b61b36419 100644
--- a/include/linux/phy/phy.h
+++ b/include/linux/phy/phy.h
@@ -33,6 +33,8 @@ struct phy;
 struct phy_ops {
 	int	(*init)(struct phy *phy);
 	int	(*exit)(struct phy *phy);
+	int	(*tune)(struct phy *phy, int phy_state);
+	int	(*set)(struct phy *phy, int option, void *info);
 	int	(*power_on)(struct phy *phy);
 	int	(*power_off)(struct phy *phy);
 	struct module *owner;
@@ -137,6 +139,8 @@ void phy_pm_runtime_allow(struct phy *phy);
 void phy_pm_runtime_forbid(struct phy *phy);
 int phy_init(struct phy *phy);
 int phy_exit(struct phy *phy);
+int phy_tune(struct phy *phy, int phy_state);
+int phy_set(struct phy *phy, int option, void *info);
 int phy_power_on(struct phy *phy);
 int phy_power_off(struct phy *phy);
 static inline int phy_get_bus_width(struct phy *phy)
@@ -227,6 +231,20 @@ static inline int phy_exit(struct phy *phy)
 	return -ENOSYS;
 }
 
+static inline int phy_tune(struct phy *phy, int phy_state)
+{
+	if (!phy)
+		return 0;
+	return -ENOSYS;
+}
+
+static inline int phy_set(struct phy *phy, int option, void *info)
+{
+	if (!phy)
+		return 0;
+	return -ENOSYS;
+}
+
 static inline int phy_power_on(struct phy *phy)
 {
 	if (!phy)
