diff --git a/arch/arm64/include/asm/esr.h b/arch/arm64/include/asm/esr.h
index f1d9b6261f55..9f664ffa36bf 100644
--- a/arch/arm64/include/asm/esr.h
+++ b/arch/arm64/include/asm/esr.h
@@ -34,8 +34,13 @@
 #define ESR_EL1_EC_CP14_64	(0x0C)
 #define ESR_EL1_EC_ILL_ISS	(0x0E)
 #define ESR_EL1_EC_SVC32	(0x11)
+#define ESR_EL1_EC_HVC32	(0x12)
+#define ESR_EL1_EC_SMC32	(0x13)
 #define ESR_EL1_EC_SVC64	(0x15)
+#define ESR_EL1_EC_HVC64	(0x16)
+#define ESR_EL1_EC_SMC64	(0x17)
 #define ESR_EL1_EC_SYS64	(0x18)
+#define ESR_EL1_EC_IMP_DEF	(0x1f)
 #define ESR_EL1_EC_IABT_EL0	(0x20)
 #define ESR_EL1_EC_IABT_EL1	(0x21)
 #define ESR_EL1_EC_PC_ALIGN	(0x22)
@@ -52,6 +57,7 @@
 #define ESR_EL1_EC_WATCHPT_EL0	(0x34)
 #define ESR_EL1_EC_WATCHPT_EL1	(0x35)
 #define ESR_EL1_EC_BKPT32	(0x38)
+#define ESR_EL1_EC_VECTOR32	(0x3A)
 #define ESR_EL1_EC_BRK64	(0x3C)
 
 #define ESR_ELx_EC_UNKNOWN	(0x00)
@@ -134,6 +140,16 @@
 #define ESR_ELx_COND_MASK	(UL(0xF) << ESR_ELx_COND_SHIFT)
 #define ESR_ELx_WFx_ISS_WFE	(UL(1) << 0)
 
+#ifndef __ASSEMBLY__
+#include <asm/types.h>
+
+const char *esr_get_class_string(u32 esr);
+#endif /* __ASSEMBLY */
+
+/* The following definitions are ported back from kernel 4.4 for esr_get_class_string */
+
+#define ESR_EL1_EC_MAX		(0x3F)
+
 /* ISS field definitions for System instruction traps */
 #define ESR_ELx_SYS64_ISS_RES0_SHIFT	22
 #define ESR_ELx_SYS64_ISS_RES0_MASK	(UL(0x7) << ESR_ELx_SYS64_ISS_RES0_SHIFT)
