diff --git a/arch/arm64/include/asm/psci.h b/arch/arm64/include/asm/psci.h
index e5312ea0ec1a..7a54c1cf9d67 100644
--- a/arch/arm64/include/asm/psci.h
+++ b/arch/arm64/include/asm/psci.h
@@ -16,4 +16,11 @@
 
 int psci_init(void);
 
+#define PSCI_UNUSED_INDEX		128
+#define PSCI_CLUSTER_SLEEP		(PSCI_UNUSED_INDEX)
+#define PSCI_SYSTEM_IDLE		(PSCI_UNUSED_INDEX + 1)
+#define PSCI_SYSTEM_IDLE_CLUSTER_SLEEP	(PSCI_UNUSED_INDEX + 2)
+#define PSCI_SYSTEM_SLEEP		(PSCI_UNUSED_INDEX + 3)
+#define PSCI_SYSTEM_CP_CALL		(PSCI_UNUSED_INDEX + 4)
+
 #endif /* __ASM_PSCI_H */
