diff --git a/drivers/hid/usbhid/usbhid.h b/drivers/hid/usbhid/usbhid.h
index f633c24ce28b..a6bcbbd6b7fa 100644
--- a/drivers/hid/usbhid/usbhid.h
+++ b/drivers/hid/usbhid/usbhid.h
@@ -92,6 +92,9 @@ struct usbhid_device {
 	unsigned int retry_delay;                                       /* Delay length in ms */
 	struct work_struct reset_work;                                  /* Task context for resets */
 	wait_queue_head_t wait;						/* For sleeping */
+#ifdef CONFIG_USB_DEBUG_DETAILED_LOG
+	unsigned long in_err_isr;
+#endif
 };
 
 #define	hid_to_usb_dev(hid_dev) \
