diff --git a/include/uapi/asm-generic/socket.h b/include/uapi/asm-generic/socket.h
index ea0796bdcf88..7ab1a9dc36df 100644
--- a/include/uapi/asm-generic/socket.h
+++ b/include/uapi/asm-generic/socket.h
@@ -82,4 +82,12 @@
 
 #define SO_BPF_EXTENSIONS	48
 
+/* START_OF_KNOX_NPA */
+#define SO_SET_DOMAIN_NAME	55
+#define SO_SET_DNS_UID		56
+/* END_OF_KNOX_NPA */
+
+/* START_OF_KNOX_NPA */
+#define SO_SET_DNS_PID		58
+/* END_OF_KNOX_NPA */
 #endif /* __ASM_GENERIC_SOCKET_H */
