diff --git a/include/linux/mmc/core.h b/include/linux/mmc/core.h
index 873ec9deb055..ad775fd6cd2c 100644
--- a/include/linux/mmc/core.h
+++ b/include/linux/mmc/core.h
@@ -95,7 +95,9 @@ struct mmc_command {
  *              actively failing requests
  */
 
+	unsigned int		cmd_timeout_ms;	/* in milliseconds */
 	unsigned int		busy_timeout;	/* busy detect timeout in ms */
+
 	/* Set this flag only for blocking sanitize request */
 	bool			sanitize_busy;
 
@@ -192,6 +194,7 @@ extern void mmc_set_data_timeout(struct mmc_data *, const struct mmc_card *);
 extern unsigned int mmc_align_data_size(struct mmc_card *, unsigned int);
 
 extern int __mmc_claim_host(struct mmc_host *host, atomic_t *abort);
+extern int mmc_try_claim_host(struct mmc_host *host, unsigned int delay);
 extern void mmc_release_host(struct mmc_host *host);
 
 extern void mmc_get_card(struct mmc_card *card);
