diff --git a/include/linux/mmc/mmc.h b/include/linux/mmc/mmc.h
index 9007d262ec41..906658346433 100644
--- a/include/linux/mmc/mmc.h
+++ b/include/linux/mmc/mmc.h
@@ -301,6 +301,7 @@ struct _mmc_csd {
 #define EXT_CSD_PART_CONFIG		179	/* R/W */
 #define EXT_CSD_ERASED_MEM_CONT		181	/* RO */
 #define EXT_CSD_BUS_WIDTH		183	/* R/W */
+#define EXT_CSD_STORBE_SUPPORT		184	/* R/W */
 #define EXT_CSD_HS_TIMING		185	/* R/W */
 #define EXT_CSD_POWER_CLASS		187	/* R/W */
 #define EXT_CSD_REV			192	/* RO */
@@ -332,9 +333,22 @@ struct _mmc_csd {
 #define EXT_CSD_GENERIC_CMD6_TIME	248	/* RO */
 #define EXT_CSD_CACHE_SIZE		249	/* RO, 4 bytes */
 #define EXT_CSD_PWR_CL_DDR_200_360	253	/* RO */
-#define EXT_CSD_PRE_EOL_INFO		267	/* RO */
-#define EXT_CSD_DEVICE_LIFE_TIME_EST_TYP_A	268	/* RO */
-#define EXT_CSD_DEVICE_LIFE_TIME_EST_TYP_B	269	/* RO */
+/* additional : eMMC v5.0 or later Only */
+#define EXT_CSD_DEVICE_LIFE_TIME_EST_TYPE_B	269	/* RO */
+#define EXT_CSD_DEVICE_LIFE_TIME_EST_TYPE_A	268	/* RO */
+#define EXT_CSD_PRE_EOL_INFO			267	/* RO */
+#define EXT_CSD_OPTIMAL_TRIM_UNIT_SIZE		264	/* RO */
+#define EXT_CSD_DEVICE_VERSION			262	/* RO, 2Byte */
+#define EXT_CSD_FIRMWARE_VERSION		254	/* RO, 8Byte */
+
+/* additional : eMMC moviNAND VMX(PRE eMMC v5.0) device Only */
+#define EXT_CSD_PREv5_FIRMWARE_VERSION		259	/* RO */
+#define EXT_CSD_PREv5_CTRL_VERSION		258	/* RO */
+#define EXT_CSD_PREv5_OPT_ERASE_SIZE		257	/* RO */
+#define EXT_CSD_PREv5_OPT_WRITE_SIZE		256	/* RO */
+#define EXT_CSD_PREv5_PRE_EOL_INFO		255	/* RO */
+#define EXT_CSD_PREv5_LIFE_TIME_EST		254	/* RO */
+
 #define EXT_CSD_TAG_UNIT_SIZE		498	/* RO */
 #define EXT_CSD_DATA_TAG_SUPPORT	499	/* RO */
 #define EXT_CSD_MAX_PACKED_WRITES	500	/* RO */
@@ -390,6 +404,7 @@ struct _mmc_csd {
 #define EXT_CSD_BUS_WIDTH_8	2	/* Card is in 8 bit mode */
 #define EXT_CSD_DDR_BUS_WIDTH_4	5	/* Card is in 4 bit DDR mode */
 #define EXT_CSD_DDR_BUS_WIDTH_8	6	/* Card is in 8 bit DDR mode */
+#define EXT_CSD_STROBE_ENHANCED_EN (1<<7) /* Card in enhanced strobe mode */
 
 #define EXT_CSD_TIMING_BC	0	/* Backwards compatility */
 #define EXT_CSD_TIMING_HS	1	/* High speed */
