diff --git a/fs/direct-io.c b/fs/direct-io.c
index a325d8c5ec2d..92a309c5898e 100644
--- a/fs/direct-io.c
+++ b/fs/direct-io.c
@@ -395,11 +395,13 @@ static inline void dio_bio_submit(struct dio *dio, struct dio_submit *sdio)
 	if (dio->is_async && dio->rw == READ)
 		bio_set_pages_dirty(bio);
 
+	dio->rw = (dio->rw == READ) ? KERNEL_READ : KERNEL_WRITE;
 	if (sdio->submit_io)
 		sdio->submit_io(dio->rw, bio, dio->inode,
 			       sdio->logical_offset_in_bio);
 	else
 		submit_bio(dio->rw, bio);
+	dio->rw = (dio->rw == KERNEL_READ) ? READ : WRITE;
 
 	sdio->bio = NULL;
 	sdio->boundary = 0;
