diff --git a/fs/internal.h b/fs/internal.h
index b1f7c4a90321..0f2fc119b100 100644
--- a/fs/internal.h
+++ b/fs/internal.h
@@ -44,6 +44,14 @@ extern void guard_bio_eod(int rw, struct bio *bio);
  */
 extern void __init chrdev_init(void);
 
+#ifdef CONFIG_PROC_DLOG
+/*
+ * dlog_hook.c
+ */
+void dlog_hook(struct dentry *, struct inode *, struct path *);
+void dlog_hook_rmdir(struct dentry *, struct path *);
+#endif
+
 /*
  * namei.c
  */
