diff --git a/fs/proc_namespace.c b/fs/proc_namespace.c
index 24fa4305f269..f16393a45210 100644
--- a/fs/proc_namespace.c
+++ b/fs/proc_namespace.c
@@ -114,7 +114,7 @@ static int show_vfsmnt(struct seq_file *m, struct vfsmount *mnt)
 		goto out;
 	show_mnt_opts(m, mnt);
 	if (sb->s_op->show_options2)
-			err = sb->s_op->show_options2(mnt, m, mnt_path.dentry);
+		err = sb->s_op->show_options2(mnt, m, mnt_path.dentry);
 	else if (sb->s_op->show_options)
 		err = sb->s_op->show_options(m, mnt_path.dentry);
 	seq_puts(m, " 0 0\n");
@@ -176,9 +176,9 @@ static int show_mountinfo(struct seq_file *m, struct vfsmount *mnt)
 	err = show_sb_opts(m, sb);
 	if (err)
 		goto out;
-	if (sb->s_op->show_options2) {
+	if (sb->s_op->show_options2)
 		err = sb->s_op->show_options2(mnt, m, mnt->mnt_root);
-	} else if (sb->s_op->show_options)
+	else if (sb->s_op->show_options)
 		err = sb->s_op->show_options(m, mnt->mnt_root);
 	seq_putc(m, '\n');
 out:
