diff --git a/drivers/staging/android/uapi/binder.h b/drivers/staging/android/uapi/binder.h
index 046aac6727ef..6094dd8c00a0 100644
--- a/drivers/staging/android/uapi/binder.h
+++ b/drivers/staging/android/uapi/binder.h
@@ -121,7 +121,7 @@ struct binder_object_header {
  */
 struct flat_binder_object {
 	struct binder_object_header	hdr;
-	__u32				flags;
+	__u32		flags;
 
 	/* 8 bytes of data. */
 	union {
@@ -272,6 +272,7 @@ struct binder_node_info_for_ref {
 #define BINDER_GET_NODE_DEBUG_INFO	_IOWR('b', 11, struct binder_node_debug_info)
 #define BINDER_GET_NODE_INFO_FOR_REF	_IOWR('b', 12, struct binder_node_info_for_ref)
 #define BINDER_SET_CONTEXT_MGR_EXT	_IOW('b', 13, struct flat_binder_object)
+#define BINDER_SET_SYSTEM_SERVER_PID		_IOW('b', 14, __u32)
 
 /*
  * NOTE: Two special error codes you should check for when calling
@@ -369,7 +370,6 @@ enum binder_driver_return_protocol {
 
 	BR_OK = _IO('r', 1),
 	/* No parameters! */
-
 	BR_TRANSACTION_SEC_CTX = _IOR('r', 2,
 				      struct binder_transaction_data_secctx),
 	/*
