diff --git a/drivers/staging/android/uapi/ion.h b/drivers/staging/android/uapi/ion.h
index 6aa495673370..e6345832f60d 100644
--- a/drivers/staging/android/uapi/ion.h
+++ b/drivers/staging/android/uapi/ion.h
@@ -63,7 +63,15 @@ enum ion_heap_type {
 #define ION_FLAG_CACHED_NEEDS_SYNC 2	/* mappings of this buffer will created
 					   at mmap time, if this is set
 					   caches must be managed manually */
-
+#define ION_FLAG_NOZEROED 8		/* Allocated buffer is not initialized
+					   with zero value and userspace is not
+					   able to access the buffer
+					 */
+#define ION_FLAG_PROTECTED 16		/* this buffer would be used in secure
+					   world. if this is set, all cpu accesses
+					   are prohibited.
+					 */
+#define ION_FLAG_SYNC_FORCE 32		/* cache sync forcely at allocation */
 /**
  * DOC: Ion Userspace API
  *
@@ -106,6 +114,13 @@ struct ion_fd_data {
 	int fd;
 };
 
+struct ion_fd_partial_data {
+	ion_user_handle_t handle;
+	int fd;
+	off_t offset;
+	size_t len;
+};
+
 /**
  * struct ion_handle_data - a handle passed to/from the kernel
  * @handle:	a handle
@@ -127,6 +142,27 @@ struct ion_custom_data {
 	unsigned long arg;
 };
 
+/**
+ * struct ion_preload_data - metadata for preload buffers
+ * @heap_id_mask:	mask of heap ids to allocate from
+ * @len:		size of the allocation
+ * @flags:		flags passed to heap
+ * @count:		number of buffers of the allocation
+ *
+ * Provided by userspace as an argument to the ioctl
+ */
+struct ion_preload_object {
+	size_t len;
+	unsigned int count;
+};
+
+struct ion_preload_data {
+	unsigned int heap_id_mask;
+	unsigned int flags;
+	unsigned int count;
+	struct ion_preload_object *obj;
+};
+
 #define ION_IOC_MAGIC		'I'
 
 /**
@@ -184,6 +220,12 @@ struct ion_custom_data {
  * this will make the buffer in memory coherent.
  */
 #define ION_IOC_SYNC		_IOWR(ION_IOC_MAGIC, 7, struct ion_fd_data)
+#define ION_IOC_SYNC_PARTIAL	_IOWR(ION_IOC_MAGIC, 9, struct ion_fd_partial_data)
+
+/**
+ * DOC: ION_IOC_PRELOAD_ALLOC - prefetches pages to page pool
+ */
+#define ION_IOC_PRELOAD_ALLOC	_IOW(ION_IOC_MAGIC, 8, struct ion_preload_data)
 
 /**
  * DOC: ION_IOC_CUSTOM - call architecture specific ion ioctl
