diff --git a/include/uapi/linux/fcntl.h b/include/uapi/linux/fcntl.h
index beed138bd359..7e3ba2531dda 100644
--- a/include/uapi/linux/fcntl.h
+++ b/include/uapi/linux/fcntl.h
@@ -27,6 +27,17 @@
 #define F_SETPIPE_SZ	(F_LINUX_SPECIFIC_BASE + 7)
 #define F_GETPIPE_SZ	(F_LINUX_SPECIFIC_BASE + 8)
 
+#ifdef CONFIG_FIVE
+#define F_FIVE_SIGN	(F_LINUX_SPECIFIC_BASE + 100)
+#define F_FIVE_VERIFY_ASYNC	(F_LINUX_SPECIFIC_BASE + 101)
+#define F_FIVE_VERIFY_SYNC	(F_LINUX_SPECIFIC_BASE + 102)
+#if defined(CONFIG_FIVE_PA_FEATURE) || defined(CONFIG_PROCA)
+#define F_FIVE_PA_SETXATTR	(F_LINUX_SPECIFIC_BASE + 103)
+#endif
+#define F_FIVE_EDIT		(F_LINUX_SPECIFIC_BASE + 104)
+#define F_FIVE_CLOSE		(F_LINUX_SPECIFIC_BASE + 105)
+#endif
+
 /*
  * Set/Get seals
  */
