diff --git a/include/uapi/linux/magic.h b/include/uapi/linux/magic.h
index fc8c9a4d1fd8..0aa6d0186f7c 100644
--- a/include/uapi/linux/magic.h
+++ b/include/uapi/linux/magic.h
@@ -53,6 +53,7 @@
 #define REISER2FS_JR_SUPER_MAGIC_STRING	"ReIsEr3Fs"
 
 #define SDCARDFS_SUPER_MAGIC	0x5dca2df5
+#define SDFAT_SUPER_MAGIC       0x5EC5DFA4
 
 #define SMB_SUPER_MAGIC		0x517B
 #define CGROUP_SUPER_MAGIC	0x27e0eb
