diff --git a/drivers/usb/gadget/legacy/acm_ms.c b/drivers/usb/gadget/legacy/acm_ms.c
index 654f11ee9be9..c30b7b572465 100644
--- a/drivers/usb/gadget/legacy/acm_ms.c
+++ b/drivers/usb/gadget/legacy/acm_ms.c
@@ -147,6 +147,10 @@ static int __init acm_ms_do_config(struct usb_configuration *c)
 	if (status < 0)
 		goto put_msg;
 
+	status = fsg_common_run_thread(opts->common);
+	if (status)
+		goto remove_acm;
+
 	status = usb_add_function(c, f_msg);
 	if (status)
 		goto remove_acm;
