diff --git a/drivers/scsi/hosts.c b/drivers/scsi/hosts.c
index 6de80e352871..fd72c4f03671 100644
--- a/drivers/scsi/hosts.c
+++ b/drivers/scsi/hosts.c
@@ -245,13 +245,10 @@ int scsi_add_host_with_dma(struct Scsi_Host *shost, struct device *dev,
 
 	pm_runtime_set_active(&shost->shost_gendev);
 	pm_runtime_enable(&shost->shost_gendev);
-	device_enable_async_suspend(&shost->shost_gendev);
 
 	scsi_host_set_state(shost, SHOST_RUNNING);
 	get_device(shost->shost_gendev.parent);
 
-	device_enable_async_suspend(&shost->shost_dev);
-
 	error = device_add(&shost->shost_dev);
 	if (error)
 		goto out_del_gendev;
