diff --git a/drivers/staging/comedi/drivers/dt3000.c b/drivers/staging/comedi/drivers/dt3000.c
index b0623bb40acb..825561046b6f 100644
--- a/drivers/staging/comedi/drivers/dt3000.c
+++ b/drivers/staging/comedi/drivers/dt3000.c
@@ -370,9 +370,9 @@ static irqreturn_t dt3k_interrupt(int irq, void *d)
 static int dt3k_ns_to_timer(unsigned int timer_base, unsigned int *nanosec,
 			    unsigned int flags)
 {
-	unsigned int divider, base, prescale;
+	int divider, base, prescale;
 
-	/* This function needs improvement */
+	/* This function needs improvment */
 	/* Don't know if divider==0 works. */
 
 	for (prescale = 0; prescale < 16; prescale++) {
@@ -386,7 +386,7 @@ static int dt3k_ns_to_timer(unsigned int timer_base, unsigned int *nanosec,
 			divider = (*nanosec) / base;
 			break;
 		case CMDF_ROUND_UP:
-			divider = DIV_ROUND_UP(*nanosec, base);
+			divider = (*nanosec) / base;
 			break;
 		}
 		if (divider < 65536) {
@@ -396,7 +396,7 @@ static int dt3k_ns_to_timer(unsigned int timer_base, unsigned int *nanosec,
 	}
 
 	prescale = 15;
-	base = timer_base * (prescale + 1);
+	base = timer_base * (1 << prescale);
 	divider = 65535;
 	*nanosec = divider * base;
 	return (prescale << 16) | (divider);
