diff --git a/net/netlink/af_netlink.c b/net/netlink/af_netlink.c
index 029dec900748..598c8697151e 100644
--- a/net/netlink/af_netlink.c
+++ b/net/netlink/af_netlink.c
@@ -1718,6 +1718,9 @@ out:
 	return err;
 }
 
+/* FIXME: will be removed, debugging code for P160223-00802 */
+extern void *memchr_inv(const void *start, int c, size_t bytes);
+
 static int netlink_recvmsg(struct kiocb *kiocb, struct socket *sock,
 			   struct msghdr *msg, size_t len,
 			   int flags)
@@ -1744,6 +1747,16 @@ static int netlink_recvmsg(struct kiocb *kiocb, struct socket *sock,
 
 #ifdef CONFIG_COMPAT_NETLINK_MESSAGES
 	if (unlikely(skb_shinfo(skb)->frag_list)) {
+		/* FIXME: will be removed, debugging code for P160223-00802 */
+		{
+			char *tmp = (char *)skb_shinfo(skb);
+			if (memchr_inv(tmp, 0x6b, 8) == NULL) {
+				pr_err("POISON_FREE: data_skb:0x%p, data_skb->head:0x%p\n",
+					data_skb, data_skb->head);
+				BUG();
+			}
+		}
+
 		/*
 		 * If this skb has a frag_list, then here that means that we
 		 * will have to use the frag_list skb's data for compat tasks
