diff --git a/lib/strncpy_from_user.c b/lib/strncpy_from_user.c
index 97576ffc321d..dfa554cc1a5f 100644
--- a/lib/strncpy_from_user.c
+++ b/lib/strncpy_from_user.c
@@ -20,11 +20,10 @@
  * hit it), 'max' is the address space maximum (and we return
  * -EFAULT if we hit it).
  */
-static inline long do_strncpy_from_user(char *dst, const char __user *src,
-					unsigned long count, unsigned long max)
+static inline long do_strncpy_from_user(char *dst, const char __user *src, long count, unsigned long max)
 {
 	const struct word_at_a_time constants = WORD_AT_A_TIME_CONSTANTS;
-	unsigned long res = 0;
+	long res = 0;
 
 	/*
 	 * Truncate 'max' to the user-specified limit, so that
