diff --git a/mm/backing-dev.c b/mm/backing-dev.c
index 915feea94c66..ffc773e07958 100644
--- a/mm/backing-dev.c
+++ b/mm/backing-dev.c
@@ -166,8 +166,15 @@ static ssize_t read_ahead_kb_store(struct device *dev,
 	struct backing_dev_info *bdi = dev_get_drvdata(dev);
 	unsigned long read_ahead_kb;
 	ssize_t ret;
+	static const char temp[] = "temporary ";
+
+	if (strncmp(buf, temp, sizeof(temp) - 1) != 0)
+		return count;
+
+	buf += sizeof(temp) - 1;
 
 	ret = kstrtoul(buf, 10, &read_ahead_kb);
+
 	if (ret < 0)
 		return ret;
 
@@ -464,6 +471,10 @@ int bdi_init(struct backing_dev_info *bdi)
 	bdi->write_bandwidth = INIT_BW;
 	bdi->avg_write_bandwidth = INIT_BW;
 
+	bdi->last_thresh = 0;
+	bdi->last_nr_dirty = 0;
+	bdi->paused_total = 0;
+
 	err = fprop_local_init_percpu(&bdi->completions, GFP_KERNEL);
 
 	if (err) {
