diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index 926ae8487c63..abaab256e7a5 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -1514,11 +1514,15 @@ static unsigned long mem_cgroup_margin(struct mem_cgroup *memcg)
 
 int mem_cgroup_swappiness(struct mem_cgroup *memcg)
 {
+#if defined(CONFIG_MEMCG_FORCE_USE_VM_SWAPPINESS)
+	return vm_swappiness;
+#else
 	/* root ? */
 	if (mem_cgroup_disabled() || !memcg->css.parent)
 		return vm_swappiness;
 
 	return memcg->swappiness;
+#endif
 }
 
 /*
@@ -4469,7 +4473,17 @@ static int memcg_stat_show(struct seq_file *m, void *v)
 
 	return 0;
 }
+static u64 mem_cgroup_vmpressure_read(struct cgroup_subsys_state *css,
+				      struct cftype *cft)
+{
+	struct mem_cgroup *memcg = mem_cgroup_from_css(css);
+	struct vmpressure *vmpr = memcg_to_vmpressure(memcg);
+	unsigned long vmpressure;
+
+	vmpressure = vmpr->pressure;
 
+	return vmpressure;
+}
 static u64 mem_cgroup_swappiness_read(struct cgroup_subsys_state *css,
 				      struct cftype *cft)
 {
@@ -4483,7 +4497,11 @@ static int mem_cgroup_swappiness_write(struct cgroup_subsys_state *css,
 {
 	struct mem_cgroup *memcg = mem_cgroup_from_css(css);
 
-	if (val > 100)
+#if CONFIG_MEMCG_HIGHER_SWAPPINESS
+	if ((val > 200) || ((val > 200) && !css->parent))
+#else
+	if ((val > 200) || ((val > 100) && !css->parent))
+#endif
 		return -EINVAL;
 
 	if (css->parent)
@@ -5202,6 +5220,10 @@ static struct cftype mem_cgroup_files[] = {
 	{
 		.name = "pressure_level",
 	},
+	{
+		.name = "vmpressure",
+		.read_u64 = mem_cgroup_vmpressure_read,
+	},
 #ifdef CONFIG_NUMA
 	{
 		.name = "numa_stat",
