diff --git a/mm/process_vm_access.c b/mm/process_vm_access.c
index b2dfa8cf6e4c..f08a9390592e 100644
--- a/mm/process_vm_access.c
+++ b/mm/process_vm_access.c
@@ -16,6 +16,7 @@
 #include <linux/ptrace.h>
 #include <linux/slab.h>
 #include <linux/syscalls.h>
+#include <linux/task_integrity.h>
 
 #ifdef CONFIG_COMPAT
 #include <linux/compat.h>
@@ -197,7 +198,7 @@ static ssize_t process_vm_rw_core(pid_t pid, struct iov_iter *iter,
 		goto free_proc_pages;
 	}
 
-	mm = mm_access(task, PTRACE_MODE_ATTACH_REALCREDS);
+	mm = mm_access(task, PTRACE_MODE_ATTACH);
 	if (!mm || IS_ERR(mm)) {
 		rc = IS_ERR(mm) ? PTR_ERR(mm) : -ESRCH;
 		/*
@@ -209,6 +210,10 @@ static ssize_t process_vm_rw_core(pid_t pid, struct iov_iter *iter,
 		goto put_task_struct;
 	}
 
+	rc = five_process_vm_rw(task, vm_write);
+	if (rc)
+		goto put_task_struct;
+
 	for (i = 0; i < riovcnt && iov_iter_count(iter) && !rc; i++)
 		rc = process_vm_rw_single_vec(
 			(unsigned long)rvec[i].iov_base, rvec[i].iov_len,
