diff --git a/mm/sparse-vmemmap.c b/mm/sparse-vmemmap.c
index 4cba9c2783a1..86a9a4b6cbe4 100644
--- a/mm/sparse-vmemmap.c
+++ b/mm/sparse-vmemmap.c
@@ -126,9 +126,24 @@ pmd_t * __meminit vmemmap_pmd_populate(pud_t *pud, unsigned long addr, int node)
 
 pud_t * __meminit vmemmap_pud_populate(pgd_t *pgd, unsigned long addr, int node)
 {
+#ifdef CONFIG_TIMA_RKP
+	int rkp_do = 0; 
+#endif 	
+	void *p = NULL ;
 	pud_t *pud = pud_offset(pgd, addr);
 	if (pud_none(*pud)) {
-		void *p = vmemmap_alloc_block(PAGE_SIZE, node);
+#ifdef CONFIG_TIMA_RKP
+#ifdef CONFIG_KNOX_KAP
+		if (boot_mode_security)  rkp_do = 1;
+#endif
+		if( rkp_do ){
+			p =  rkp_ro_alloc();
+		}else{
+			p = vmemmap_alloc_block(PAGE_SIZE, node);
+		}
+#else /* !CONFIG_TIMA_RKP */
+		p = vmemmap_alloc_block(PAGE_SIZE, node);
+#endif
 		if (!p)
 			return NULL;
 		pud_populate(&init_mm, pud, p);
