diff --git a/mm/swap_state.c b/mm/swap_state.c
index 652d864d3178..5c18a771b4de 100644
--- a/mm/swap_state.c
+++ b/mm/swap_state.c
@@ -392,6 +392,7 @@ struct page *read_swap_cache_async(swp_entry_t entry, gfp_t gfp_mask,
 	return found_page;
 }
 
+#ifdef CONFIG_SWAP_ENABLE_READAHEAD
 static unsigned long swapin_nr_pages(unsigned long offset)
 {
 	static unsigned long prev_offset;
@@ -435,6 +436,7 @@ static unsigned long swapin_nr_pages(unsigned long offset)
 
 	return pages;
 }
+#endif
 
 /**
  * swapin_readahead - swap in pages in hope we need them soon
@@ -458,6 +460,7 @@ static unsigned long swapin_nr_pages(unsigned long offset)
 struct page *swapin_readahead(swp_entry_t entry, gfp_t gfp_mask,
 			struct vm_area_struct *vma, unsigned long addr)
 {
+#ifdef CONFIG_SWAP_ENABLE_READAHEAD
 	struct page *page;
 	unsigned long entry_offset = swp_offset(entry);
 	unsigned long offset = entry_offset;
@@ -490,5 +493,6 @@ struct page *swapin_readahead(swp_entry_t entry, gfp_t gfp_mask,
 
 	lru_add_drain();	/* Push any new pages onto the LRU now */
 skip:
+#endif
 	return read_swap_cache_async(entry, gfp_mask, vma, addr);
 }
