diff --git a/mm/vmpressure.c b/mm/vmpressure.c
index c5afd573d7da..b54d802b181d 100644
--- a/mm/vmpressure.c
+++ b/mm/vmpressure.c
@@ -46,7 +46,7 @@ static const unsigned long vmpressure_win = SWAP_CLUSTER_MAX * 16;
  * essence, they are percents: the higher the value, the more number
  * unsuccessful reclaims there were.
  */
-static const unsigned int vmpressure_level_med = 60;
+static const unsigned int vmpressure_level_med = CONFIG_VMPRESSURE_LEVEL_MED;
 static const unsigned int vmpressure_level_critical = 95;
 
 /*
@@ -109,7 +109,7 @@ static enum vmpressure_levels vmpressure_level(unsigned long pressure)
 }
 
 static enum vmpressure_levels vmpressure_calc_level(unsigned long scanned,
-						    unsigned long reclaimed)
+						    unsigned long reclaimed, struct vmpressure *vmpr)
 {
 	unsigned long scale = scanned + reclaimed;
 	unsigned long pressure;
@@ -123,6 +123,7 @@ static enum vmpressure_levels vmpressure_calc_level(unsigned long scanned,
 	 */
 	pressure = scale - (reclaimed * scale / scanned);
 	pressure = pressure * 100 / scale;
+	vmpr->pressure = pressure;
 
 	pr_debug("%s: %3lu  (s: %lu  r: %lu)\n", __func__, pressure,
 		 scanned, reclaimed);
@@ -143,7 +144,7 @@ static bool vmpressure_event(struct vmpressure *vmpr,
 	enum vmpressure_levels level;
 	bool signalled = false;
 
-	level = vmpressure_calc_level(scanned, reclaimed);
+	level = vmpressure_calc_level(scanned, reclaimed, vmpr);
 
 	mutex_lock(&vmpr->events_lock);
 
