diff --git a/fs/cifs/file.c b/fs/cifs/file.c
index 82c3bedfa40a..c986860dbaf5 100644
--- a/fs/cifs/file.c
+++ b/fs/cifs/file.c
@@ -253,12 +253,6 @@ cifs_nt_open(char *full_path, struct inode *inode, struct cifs_sb_info *cifs_sb,
 		rc = cifs_get_inode_info(&inode, full_path, buf, inode->i_sb,
 					 xid, fid);
 
-	if (rc) {
-		server->ops->close(xid, tcon, fid);
-		if (rc == -ESTALE)
-			rc = -EOPENSTALE;
-	}
-
 out:
 	kfree(buf);
 	return rc;
