diff --git a/fs/cifs/inode.c b/fs/cifs/inode.c
index a7c1362cb09a..c5f86438a9b0 100644
--- a/fs/cifs/inode.c
+++ b/fs/cifs/inode.c
@@ -1858,11 +1858,11 @@ cifs_invalidate_mapping(struct inode *inode)
  * @word: long word containing the bit lock
  */
 static int
-cifs_wait_bit_killable(struct wait_bit_key *key)
+cifs_wait_bit_killable(struct wait_bit_key *key, int mode)
 {
-	if (fatal_signal_pending(current))
-		return -ERESTARTSYS;
 	freezable_schedule_unsafe();
+	if (signal_pending_state(mode, current))
+		return -ERESTARTSYS;
 	return 0;
 }
 
