diff --git a/fs/cifs/smb2pdu.c b/fs/cifs/smb2pdu.c
index b3f28c2205fd..fbc10a628b6b 100644
--- a/fs/cifs/smb2pdu.c
+++ b/fs/cifs/smb2pdu.c
@@ -171,7 +171,7 @@ smb2_reconnect(__le16 smb2_command, struct cifs_tcon *tcon)
 	if (tcon == NULL)
 		return 0;
 
-	if (smb2_command == SMB2_TREE_CONNECT || smb2_command == SMB2_IOCTL)
+	if (smb2_command == SMB2_TREE_CONNECT)
 		return 0;
 
 	if (tcon->tidStatus == CifsExiting) {
@@ -636,12 +636,7 @@ ssetup_ntlmssp_authenticate:
 	else
 		req->SecurityMode = 0;
 
-#ifdef CONFIG_CIFS_DFS_UPCALL
-	req->Capabilities = cpu_to_le32(SMB2_GLOBAL_CAP_DFS);
-#else
 	req->Capabilities = 0;
-#endif /* DFS_UPCALL */
-
 	req->Channel = 0; /* MBZ */
 
 	iov[0].iov_base = (char *)req;
