diff --git a/drivers/pps/pps.c b/drivers/pps/pps.c
index 76ae38450aea..2f07cd615665 100644
--- a/drivers/pps/pps.c
+++ b/drivers/pps/pps.c
@@ -129,14 +129,6 @@ static long pps_cdev_ioctl(struct file *file,
 			pps->params.mode |= PPS_CANWAIT;
 		pps->params.api_version = PPS_API_VERS;
 
-		/*
-		 * Clear unused fields of pps_kparams to avoid leaking
-		 * uninitialized data of the PPS_SETPARAMS caller via
-		 * PPS_GETPARAMS
-		 */
-		pps->params.assert_off_tu.flags = 0;
-		pps->params.clear_off_tu.flags = 0;
-
 		spin_unlock_irq(&pps->lock);
 
 		break;
