diff --git a/fs/nfs/internal.h b/fs/nfs/internal.h
index e1acc1c4ce65..b75d32f37335 100644
--- a/fs/nfs/internal.h
+++ b/fs/nfs/internal.h
@@ -349,7 +349,7 @@ extern int nfs_drop_inode(struct inode *);
 extern void nfs_clear_inode(struct inode *);
 extern void nfs_evict_inode(struct inode *);
 void nfs_zap_acl_cache(struct inode *inode);
-extern int nfs_wait_bit_killable(struct wait_bit_key *key);
+extern int nfs_wait_bit_killable(struct wait_bit_key *key, int mode);
 
 /* super.c */
 extern const struct super_operations nfs_sops;
