diff --git a/security/keys/request_key.c b/security/keys/request_key.c
index 48b9df73038d..2a60a75ea313 100644
--- a/security/keys/request_key.c
+++ b/security/keys/request_key.c
@@ -463,13 +463,13 @@ static struct key *construct_key_and_link(struct keyring_search_context *ctx,
 
 	kenter("");
 
+	if (ctx->index_key.type == &key_type_keyring)
+		return ERR_PTR(-EPERM);
+
 	ret = construct_get_dest_keyring(&dest_keyring);
 	if (ret)
 		goto error;
 
-	if (ctx->index_key.type == &key_type_keyring)
-		return ERR_PTR(-EPERM);
-
 	user = key_user_lookup(current_fsuid());
 	if (!user) {
 		ret = -ENOMEM;
