diff --git a/security/keys/user_defined.c b/security/keys/user_defined.c
index 7cf22260bdff..6ce8a92451e6 100644
--- a/security/keys/user_defined.c
+++ b/security/keys/user_defined.c
@@ -163,6 +163,12 @@ void user_destroy(struct key *key)
 {
 	struct user_key_payload *upayload = key->payload.data;
 
+#ifdef CONFIG_CRYPTO_FIPS
+	if(upayload)
+	{
+		memset(upayload->data, 0, upayload->datalen);
+	}
+#endif
 	kfree(upayload);
 }
 
