diff --git a/include/linux/amba/pl330.h b/include/linux/amba/pl330.h
index fe93758e8403..c95df58e0e7a 100644
--- a/include/linux/amba/pl330.h
+++ b/include/linux/amba/pl330.h
@@ -32,4 +32,6 @@ struct dma_pl330_platdata {
 };
 
 extern bool pl330_filter(struct dma_chan *chan, void *param);
+extern int pl330_dma_getposition(struct dma_chan *chan,
+		dma_addr_t *src, dma_addr_t *dst);
 #endif	/* __AMBA_PL330_H_ */
