diff --git a/net/sunrpc/sched.c b/net/sunrpc/sched.c
index 453edae6e105..d1ae7167db00 100644
--- a/net/sunrpc/sched.c
+++ b/net/sunrpc/sched.c
@@ -259,11 +259,11 @@ void rpc_destroy_wait_queue(struct rpc_wait_queue *queue)
 }
 EXPORT_SYMBOL_GPL(rpc_destroy_wait_queue);
 
-static int rpc_wait_bit_killable(struct wait_bit_key *key)
+static int rpc_wait_bit_killable(struct wait_bit_key *key, int mode)
 {
-	if (fatal_signal_pending(current))
-		return -ERESTARTSYS;
 	freezable_schedule_unsafe();
+	if (signal_pending_state(mode, current))
+		return -ERESTARTSYS;
 	return 0;
 }
 
