diff --git a/include/scsi/scsi_cmnd.h b/include/scsi/scsi_cmnd.h
index 522a5f27f553..5ec1823f1026 100644
--- a/include/scsi/scsi_cmnd.h
+++ b/include/scsi/scsi_cmnd.h
@@ -163,6 +163,7 @@ extern int scsi_init_io(struct scsi_cmnd *cmd, gfp_t gfp_mask);
 
 extern int scsi_dma_map(struct scsi_cmnd *cmd);
 extern void scsi_dma_unmap(struct scsi_cmnd *cmd);
+extern void scsi_dma_set_skip_cpu_sync(void);
 
 static inline unsigned scsi_sg_count(struct scsi_cmnd *cmd)
 {
