diff --git a/drivers/thermal/platform_tables.h b/drivers/thermal/platform_tables.h
new file mode 100644
index 000000000000..70a23bb709ac
--- /dev/null
+++ b/drivers/thermal/platform_tables.h
@@ -0,0 +1,14 @@
+#ifndef _DRIVERS_THERMAL_PLATFORM_H
+#define _DRIVERS_THERMAL_PLATFORM_H
+
+struct coefficients {
+	int frequency;
+	int power;
+};
+
+#define CPUFREQ_LEVEL_INDEX_END		L24
+
+struct coefficients little_cpu_coeffs[CPUFREQ_LEVEL_INDEX_END];
+struct coefficients big_cpu_coeffs[CPUFREQ_LEVEL_INDEX_END];
+
+#endif /* _DRIVERS_THERMAL_PLATFORM_H */
