diff --git a/drivers/thermal/step_wise.c b/drivers/thermal/step_wise.c
index eed5b448fbf9..ae7f8dcfdf6b 100644
--- a/drivers/thermal/step_wise.c
+++ b/drivers/thermal/step_wise.c
@@ -208,6 +208,9 @@ static int step_wise_throttle(struct thermal_zone_device *tz, int trip)
 	list_for_each_entry(instance, &tz->thermal_instances, tz_node)
 		thermal_cdev_update(instance->cdev);
 
+	if (tz->ops->throttle_cpu_hotplug)
+		tz->ops->throttle_cpu_hotplug(tz);
+
 	mutex_unlock(&tz->lock);
 
 	return 0;
