diff --git a/drivers/usb/image/microtek.c b/drivers/usb/image/microtek.c
index 65e7ab35cf07..37b44b04a701 100644
--- a/drivers/usb/image/microtek.c
+++ b/drivers/usb/image/microtek.c
@@ -727,10 +727,6 @@ static int mts_usb_probe(struct usb_interface *intf,
 
 	}
 
-	if (ep_in_current != &ep_in_set[2]) {
-		MTS_WARNING("couldn't find two input bulk endpoints. Bailing out.\n");
-		return -ENODEV;
-	}
 
 	if ( ep_out == -1 ) {
 		MTS_WARNING( "couldn't find an output bulk endpoint. Bailing out.\n" );
