diff --git a/fs/reiserfs/super.c b/fs/reiserfs/super.c
index 673568415aa9..5f4f16c98a89 100644
--- a/fs/reiserfs/super.c
+++ b/fs/reiserfs/super.c
@@ -809,6 +809,7 @@ static const struct dquot_operations reiserfs_quota_operations = {
 	.write_info = reiserfs_write_info,
 	.alloc_dquot	= dquot_alloc,
 	.destroy_dquot	= dquot_destroy,
+	.get_next_id	= dquot_get_next_id,
 };
 
 static const struct quotactl_ops reiserfs_qctl_operations = {
