diff --git a/drivers/usb/host/ehci-hub.c b/drivers/usb/host/ehci-hub.c
index ecd5d6a55ca3..10255569f2bf 100644
--- a/drivers/usb/host/ehci-hub.c
+++ b/drivers/usb/host/ehci-hub.c
@@ -579,6 +579,19 @@ static int check_reset_complete (
 			return port_status;
 		}
 
+		/* W/A for Synopsys HC HSIC port.
+		 * Return at this point to prevent port owner change
+		 * and retry port reset.
+		 */
+		if (ehci->has_synopsys_hsic_bug) {
+			if ((index + 1) == ehci->hsic_ports) {
+				ehci_err (ehci,
+					"Failed to enable HSIC port %d\n",
+					index + 1);
+				return port_status;
+			}
+		}
+
 		ehci_dbg (ehci, "port %d full speed --> companion\n",
 			index + 1);
 
@@ -1209,6 +1222,20 @@ int ehci_hub_control(
 					wIndex + 1);
 				temp |= PORT_OWNER;
 			} else {
+				ehci_vdbg (ehci, "port %d reset\n", wIndex + 1);
+
+				/* W/A for Synopsys HC HSIC port.
+				 * Disable HSIC port to prevent
+				 * the port reset failure.
+				 */
+				if (ehci->has_synopsys_hsic_bug) {
+					if ((wIndex + 1) == ehci->hsic_ports) {
+						ehci_writel(ehci,
+							temp & ~PORT_PE,
+							status_reg);
+					}
+				}
+
 				temp |= PORT_RESET;
 				temp &= ~PORT_PE;
 
