diff --git a/drivers/net/phy/national.c b/drivers/net/phy/national.c
index a63a32a35fc6..9a5f234d95b0 100644
--- a/drivers/net/phy/national.c
+++ b/drivers/net/phy/national.c
@@ -110,17 +110,14 @@ static void ns_giga_speed_fallback(struct phy_device *phydev, int mode)
 
 static void ns_10_base_t_hdx_loopack(struct phy_device *phydev, int disable)
 {
-	u16 lb_dis = BIT(1);
-
 	if (disable)
-		ns_exp_write(phydev, 0x1c0,
-			     ns_exp_read(phydev, 0x1c0) | lb_dis);
+		ns_exp_write(phydev, 0x1c0, ns_exp_read(phydev, 0x1c0) | 1);
 	else
 		ns_exp_write(phydev, 0x1c0,
-			     ns_exp_read(phydev, 0x1c0) & ~lb_dis);
+			     ns_exp_read(phydev, 0x1c0) & 0xfffe);
 
 	pr_debug("10BASE-T HDX loopback %s\n",
-		 (ns_exp_read(phydev, 0x1c0) & lb_dis) ? "off" : "on");
+		 (ns_exp_read(phydev, 0x1c0) & 0x0001) ? "off" : "on");
 }
 
 static int ns_config_init(struct phy_device *phydev)
