diff --git a/fs/fat/fat.h b/fs/fat/fat.h
index add94f15d5be..d51016741415 100644
--- a/fs/fat/fat.h
+++ b/fs/fat/fat.h
@@ -9,6 +9,13 @@
 #include <linux/mutex.h>
 #include <linux/ratelimit.h>
 #include <linux/msdos_fs.h>
+#include <linux/kobject.h>
+
+#ifdef CONFIG_FAT_SUPPORT_STLOG
+#include <linux/fslog.h>
+#else
+#define ST_LOG(fmt,...) 
+#endif
 
 /*
  * vfat shortname flags
@@ -393,6 +400,18 @@ static inline unsigned long fat_dir_hash(int logstart)
 }
 
 /* fat/misc.c */
+#ifdef CONFIG_FAT_UEVENT
+extern int fat_uevent_init(struct kset *fat_kset);
+extern void fat_uevent_uninit(void);
+extern void fat_uevent_ro_remount(struct super_block *sb);
+#else
+static inline int fat_uevent_init(struct kset *fat_kset)
+{
+	return 0;
+}
+static inline void fat_uevent_uninit(void) {};
+static inline void fat_uevent_ro_remount(struct super_block *sb) {};
+#endif
 extern __printf(3, 4) __cold
 void __fat_fs_error(struct super_block *sb, int report, const char *fmt, ...);
 #define fat_fs_error(sb, fmt, args...)		\
@@ -421,6 +440,14 @@ void fat_cache_destroy(void);
 extern const struct export_operations fat_export_ops;
 extern const struct export_operations fat_export_ops_nostale;
 
+/* fat/xattr.c */
+extern int fat_setxattr(struct dentry *dentry, const char *name,
+				const void *value, size_t size, int flags);
+extern ssize_t fat_getxattr(struct dentry *dentry, const char *name,
+				void *value, size_t size);
+extern ssize_t fat_listxattr(struct dentry *dentry, char *list, size_t size);
+extern int fat_removexattr(struct dentry *dentry, const char *name);
+
 /* helper for printk */
 typedef unsigned long long	llu;
 
