diff --git a/drivers/usb/storage/uas.c b/drivers/usb/storage/uas.c
index ebabd2294d75..72a9e001d846 100644
--- a/drivers/usb/storage/uas.c
+++ b/drivers/usb/storage/uas.c
@@ -898,14 +898,15 @@ MODULE_DEVICE_TABLE(usb, uas_usb_ids);
 static int uas_switch_interface(struct usb_device *udev,
 				struct usb_interface *intf)
 {
-	struct usb_host_interface *alt;
+	struct usb_host_interface *alt; 
 
 	alt = uas_find_uas_alt_setting(intf);
-	if (!alt)
-		return -ENODEV;
+	if (!alt)  
+		return -ENODEV;  
+
+	return usb_set_interface(udev, alt->desc.bInterfaceNumber,  
+			alt->desc.bAlternateSetting);  
 
-	return usb_set_interface(udev, alt->desc.bInterfaceNumber,
-			alt->desc.bAlternateSetting);
 }
 
 static int uas_configure_endpoints(struct uas_dev_info *devinfo)
