diff --git a/drivers/media/mtv/isdbt_port_sms3230.h b/drivers/media/mtv/isdbt_port_sms3230.h
new file mode 100755
index 000000000000..b38afb063663
--- /dev/null
+++ b/drivers/media/mtv/isdbt_port_sms3230.h
@@ -0,0 +1,36 @@
+/*
+ *
+ * drivers/media/tdmb/isdbt_port_sms3230.h
+ *
+ * isdbt driver wrapper for Siano SMS3230
+ *
+ * Copyright (C) (2015, Samsung Electronics)
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation version 2.
+ *
+ * This program is distributed "as is" WITHOUT ANY WARRANTY of any
+ * kind, whether express or implied; without even the implied warranty
+ * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ */
+#ifndef __ISDBT_PORT_H__
+#define __ISDBT_PORT_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "isdbt.h"
+#include "sms3230/smscoreapi.h"
+#include "sms3230/smsspiphy.h"
+#include "sms3230/sms3230.h"
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif
+
