diff --git a/security/selinux/netif.c b/security/selinux/netif.c
index e607b4473ef6..6374ef18656d 100644
--- a/security/selinux/netif.c
+++ b/security/selinux/netif.c
@@ -276,6 +276,12 @@ static struct notifier_block sel_netif_netdev_notifier = {
 static __init int sel_netif_init(void)
 {
 	int i;
+	
+// [ SEC_SELINUX_PORTING_COMMON
+#ifdef CONFIG_SECURITY_SELINUX_ALWAYS_ENFORCE
+	selinux_enabled = 1;
+#endif
+// ] SEC_SELINUX_PORTING_COMMON
 
 	if (!selinux_enabled)
 		return 0;
