diff --git a/crypto/ablkcipher.c b/crypto/ablkcipher.c
index cad66af5a1ae..795d1a0e3c9b 100644
--- a/crypto/ablkcipher.c
+++ b/crypto/ablkcipher.c
@@ -48,6 +48,11 @@ void __ablkcipher_walk_complete(struct ablkcipher_walk *walk)
 {
 	struct ablkcipher_buffer *p, *tmp;
 
+#ifdef CONFIG_CRYPTO_FIPS
+	if (unlikely(in_fips_err()))
+		return;
+#endif
+
 	list_for_each_entry_safe(p, tmp, &walk->buffers, entry) {
 		ablkcipher_buffer_write(p);
 		list_del(&p->entry);
@@ -105,6 +110,11 @@ int ablkcipher_walk_done(struct ablkcipher_request *req,
 	unsigned int n; /* bytes processed */
 	bool more;
 
+#ifdef CONFIG_CRYPTO_FIPS
+	if (unlikely(in_fips_err()))
+		return -EACCES;
+#endif
+
 	if (unlikely(err < 0))
 		goto finish;
 
@@ -651,6 +661,11 @@ int crypto_grab_skcipher(struct crypto_skcipher_spawn *spawn, const char *name,
 	struct crypto_alg *alg;
 	int err;
 
+#ifdef CONFIG_CRYPTO_FIPS
+	if (unlikely(in_fips_err()))
+		return -EACCES;
+#endif
+
 	type = crypto_skcipher_type(type);
 	mask = crypto_skcipher_mask(mask);
 
