diff --git a/crypto/algboss.c b/crypto/algboss.c
index 8735bc1be13f..3b3e15330edb 100644
--- a/crypto/algboss.c
+++ b/crypto/algboss.c
@@ -67,9 +67,20 @@ static int cryptomgr_probe(void *data)
 	int err;
 
 	tmpl = crypto_lookup_template(param->template);
+
+#ifndef CONFIG_CRYPTO_FIPS
 	if (!tmpl)
 		goto out;
-
+#else
+	/* change@dtl.ksingh
+	 * Below if condition needs to test for valid point
+	 * but instead it was testing for NULL. Crypto APIs never
+	 * return NULL, hence in failure case this was causing 
+	 * kernel panic
+	 */
+	if (!tmpl || IS_ERR(tmpl))
+		goto out;
+#endif
 	do {
 		if (tmpl->create) {
 			err = tmpl->create(tmpl, param->tb);
@@ -212,9 +223,10 @@ static int cryptomgr_test(void *data)
 	u32 type = param->type;
 	int err = 0;
 
-#ifdef CONFIG_CRYPTO_MANAGER_DISABLE_TESTS
+// skip test procedure if it's not from POST.
+//#ifdef CONFIG_CRYPTO_MANAGER_DISABLE_TESTS
 	goto skiptest;
-#endif
+//#endif
 
 	if (type & CRYPTO_ALG_TESTED)
 		goto skiptest;
