diff --git a/crypto/proc.c b/crypto/proc.c
index 4a0a7aad2204..f291329ac39e 100644
--- a/crypto/proc.c
+++ b/crypto/proc.c
@@ -26,8 +26,7 @@
 #ifdef CONFIG_CRYPTO_FIPS
 static struct ctl_table crypto_sysctl_table[] = {
 	{
-		.procname       = "fips_enabled",
-		.data           = &fips_enabled,
+		.procname       = "fips_status",
 		.maxlen         = sizeof(int),
 		.mode           = 0444,
 		.proc_handler   = proc_dointvec
@@ -142,11 +141,20 @@ static const struct file_operations proc_crypto_ops = {
 	.release	= seq_release
 };
 
+#ifdef CONFIG_CRYPTO_FIPS
+void crypto_init_proc(int *fips_error)
+{
+	proc_create("crypto", 0, NULL, &proc_crypto_ops);
+	crypto_sysctl_table[0].data = fips_error;
+	crypto_proc_fips_init();
+}
+#else
 void __init crypto_init_proc(void)
 {
 	proc_create("crypto", 0, NULL, &proc_crypto_ops);
 	crypto_proc_fips_init();
 }
+#endif
 
 void __exit crypto_exit_proc(void)
 {
