diff --git a/net/wireless/core.h b/net/wireless/core.h
index ddaf3b725397..a9da7261d10d 100644
--- a/net/wireless/core.h
+++ b/net/wireless/core.h
@@ -114,8 +114,10 @@ cfg80211_rdev_free_wowlan(struct cfg80211_registered_device *rdev)
 	    rdev->wiphy.wowlan_config->tcp->sock)
 		sock_release(rdev->wiphy.wowlan_config->tcp->sock);
 	kfree(rdev->wiphy.wowlan_config->tcp);
+#ifndef CONFIG_QCOM_WIFI
 	kfree(rdev->wiphy.wowlan_config);
 #endif
+#endif
 }
 
 extern struct workqueue_struct *cfg80211_wq;
