diff --git a/net/wireless/scan.c b/net/wireless/scan.c
index ef0c3b2c7e28..898e7f51bc07 100644
--- a/net/wireless/scan.c
+++ b/net/wireless/scan.c
@@ -488,8 +488,17 @@ static int cmp_bss(struct cfg80211_bss *a,
 	const u8 *ie2 = NULL;
 	int i, r;
 
+#if !(defined(CONFIG_BCM4335) || defined(CONFIG_BCM4335_MODULE) \
+        || defined(CONFIG_BCM4339) || defined(CONFIG_BCM4339_MODULE) \
+	|| defined(CONFIG_BCM43438) || defined(CONFIG_BCM43438_MODULE) \
+	|| defined(CONFIG_BCM43454) || defined(CONFIG_BCM43454_MODULE) \
+        || defined(CONFIG_BCM43455) || defined(CONFIG_BCM43455_MODULE) \
+	|| defined(CONFIG_BCM4354) || defined(CONFIG_BCM4354_MODULE) \
+        || defined(CONFIG_BCM4356) || defined(CONFIG_BCM4356_MODULE) \
+        || defined(CONFIG_BCM4358) || defined(CONFIG_BCM4358_MODULE))
 	if (a->channel != b->channel)
 		return b->channel->center_freq - a->channel->center_freq;
+#endif /* CONFIG_BCM43xx */
 
 	a_ies = rcu_access_pointer(a->ies);
 	if (!a_ies)
@@ -529,6 +538,18 @@ static int cmp_bss(struct cfg80211_bss *a,
 	if (r)
 		return r;
 
+#if (defined(CONFIG_BCM4335) || defined(CONFIG_BCM4335_MODULE) \
+        || defined(CONFIG_BCM4339) || defined(CONFIG_BCM4339_MODULE) \
+	|| defined(CONFIG_BCM43438) || defined(CONFIG_BCM43438_MODULE) \
+	|| defined(CONFIG_BCM43454) || defined(CONFIG_BCM43454_MODULE) \
+        || defined(CONFIG_BCM43455) || defined(CONFIG_BCM43455_MODULE) \
+	|| defined(CONFIG_BCM4354) || defined(CONFIG_BCM4354_MODULE) \
+        || defined(CONFIG_BCM4356) || defined(CONFIG_BCM4356_MODULE) \
+        || defined(CONFIG_BCM4358) || defined(CONFIG_BCM4358_MODULE))
+	if (a->channel != b->channel)
+		return b->channel->center_freq - a->channel->center_freq;
+#endif /* CONFIG_BCM43xx */
+
 	ie1 = cfg80211_find_ie(WLAN_EID_SSID, a_ies->data, a_ies->len);
 	ie2 = cfg80211_find_ie(WLAN_EID_SSID, b_ies->data, b_ies->len);
 
@@ -1288,6 +1309,8 @@ int cfg80211_wext_siwscan(struct net_device *dev,
 		if (wiphy->bands[i])
 			creq->rates[i] = (1 << wiphy->bands[i]->n_bitrates) - 1;
 
+	eth_broadcast_addr(creq->bssid);
+
 	rdev->scan_req = creq;
 	err = rdev_scan(rdev, creq);
 	if (err) {
