diff --git a/drivers/regulator/of_regulator.c b/drivers/regulator/of_regulator.c
index 87621db13cb7..3d2f97933c52 100644
--- a/drivers/regulator/of_regulator.c
+++ b/drivers/regulator/of_regulator.c
@@ -23,6 +23,7 @@ static void of_get_regulation_constraints(struct device_node *np,
 					struct regulator_init_data **init_data)
 {
 	const __be32 *min_uV, *max_uV;
+	const __be32 *expected_consumer;
 	struct regulation_constraints *constraints = &(*init_data)->constraints;
 	int ret;
 	u32 pval;
@@ -73,6 +74,17 @@ static void of_get_regulation_constraints(struct device_node *np,
 	ret = of_property_read_u32(np, "regulator-enable-ramp-delay", &pval);
 	if (!ret)
 		constraints->enable_time = pval;
+
+	ret = of_property_read_u32(np, "regulator-initial-mode", &pval);
+	if (!ret)
+		constraints->initial_mode = pval;
+
+	/* If expected-consumer is not null, regulator will not set lower voltage
+	 * unless all consuler list is registered.
+	 */
+	expected_consumer = of_get_property(np, "regulator-expected-consumer", NULL);
+	if (expected_consumer)
+		constraints->expected_consumer = be32_to_cpu(*expected_consumer);
 }
 
 /**
