diff --git a/net/sctp/sm_statefuns.c b/net/sctp/sm_statefuns.c
index 526be5ee3c6f..0556637e382c 100644
--- a/net/sctp/sm_statefuns.c
+++ b/net/sctp/sm_statefuns.c
@@ -3450,6 +3450,7 @@ sctp_disposition_t sctp_sf_ootb(struct net *net,
 		if (ntohs(ch->length) < sizeof(sctp_chunkhdr_t))
 			return sctp_sf_violation_chunklen(net, ep, asoc, type, arg,
 						  commands);
+						  
 
 		/* Report violation if chunk len overflows */
 		ch_end = ((__u8 *)ch) + WORD_ROUND(ntohs(ch->length));
