diff --git a/arch/arm64/boot/dts/exynos7870-j7popelte_usa_att_battery_01.dtsi b/arch/arm64/boot/dts/exynos7870-j7popelte_usa_att_battery_01.dtsi
new file mode 100755
index 000000000000..1ecd71963300
--- /dev/null
+++ b/arch/arm64/boot/dts/exynos7870-j7popelte_usa_att_battery_01.dtsi
@@ -0,0 +1,15 @@
+/*
+ * SAMSUNG UNIVERSAL5433 board device tree source
+ *
+ * Copyright (c) 2013 Samsung Electronics Co., Ltd.
+ *		http://www.samsung.com
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+*/
+
+#include "exynos7870-j7popelte-common-battery.dtsi"
+/ {
+
+};
