diff --git a/net/ipv6/ah6.c b/net/ipv6/ah6.c
index 1d7ed2fbcb0d..c6ceb862a6d8 100644
--- a/net/ipv6/ah6.c
+++ b/net/ipv6/ah6.c
@@ -665,7 +665,7 @@ static int ah6_err(struct sk_buff *skb, struct inet6_skb_parm *opt,
 
 	if (type == NDISC_REDIRECT)
 		ip6_redirect(skb, net, skb->dev->ifindex, 0,
-			     sock_net_uid(net, NULL));
+			sock_net_uid(net, NULL));
 	else
 		ip6_update_pmtu(skb, net, info, 0, 0, sock_net_uid(net, NULL));
 	xfrm_state_put(x);
