diff --git a/net/ipv6/icmp.c b/net/ipv6/icmp.c
index 9cf56041f378..0f3db92cba09 100644
--- a/net/ipv6/icmp.c
+++ b/net/ipv6/icmp.c
@@ -94,7 +94,7 @@ static void icmpv6_err(struct sk_buff *skb, struct inet6_skb_parm *opt,
 		ip6_update_pmtu(skb, net, info, 0, 0, sock_net_uid(net, NULL));
 	else if (type == NDISC_REDIRECT)
 		ip6_redirect(skb, net, skb->dev->ifindex, 0,
-			     sock_net_uid(net, NULL));
+						sock_net_uid(net, NULL));
 
 	if (!(type & ICMPV6_INFOMSG_MASK))
 		if (icmp6->icmp6_type == ICMPV6_ECHO_REQUEST)
