diff --git a/net/ipv6/ping.c b/net/ipv6/ping.c
index b0b443432682..8a64fc9011ae 100644
--- a/net/ipv6/ping.c
+++ b/net/ipv6/ping.c
@@ -232,7 +232,7 @@ static int __net_init ping_v6_proc_init_net(struct net *net)
 	return ping_proc_register(net, &ping_v6_seq_afinfo);
 }
 
-static void __net_exit ping_v6_proc_exit_net(struct net *net)
+static void __net_init ping_v6_proc_exit_net(struct net *net)
 {
 	return ping_proc_unregister(net, &ping_v6_seq_afinfo);
 }
