diff --git a/drivers/media/platform/exynos/fimc-is2/vendor/default/Makefile b/drivers/media/platform/exynos/fimc-is2/vendor/default/Makefile
new file mode 100644
index 000000000000..3f1b3858b4f0
--- /dev/null
+++ b/drivers/media/platform/exynos/fimc-is2/vendor/default/Makefile
@@ -0,0 +1,4 @@
+obj-$(CONFIG_VENDER_DEFAULT) += fimc-is-vender.o
+
+EXTRA_CFLAGS += -Idrivers/media/platform/exynos/fimc-is2
+EXTRA_CFLAGS += -Idrivers/media/platform/exynos/fimc-is2/include
