diff --git a/include/linux/usb/class-dual-role.h b/include/linux/usb/class-dual-role.h
index c6df2238012e..af42ed34944a 100644
--- a/include/linux/usb/class-dual-role.h
+++ b/include/linux/usb/class-dual-role.h
@@ -109,19 +109,18 @@ extern int dual_role_property_is_writeable(struct dual_role_phy_instance
 					   enum dual_role_property prop);
 extern void *dual_role_get_drvdata(struct dual_role_phy_instance *dual_role);
 #else /* CONFIG_DUAL_ROLE_USB_INTF */
-static inline void dual_role_instance_changed(struct dual_role_phy_instance
+static void dual_role_instance_changed(struct dual_role_phy_instance
 				       *dual_role){}
-static inline struct dual_role_phy_instance *__must_check
+static struct dual_role_phy_instance *__must_check
 devm_dual_role_instance_register(struct device *parent,
 				 const struct dual_role_phy_desc *desc)
 {
 	return ERR_PTR(-ENOSYS);
 }
-static inline void devm_dual_role_instance_unregister(struct device *dev,
+static void devm_dual_role_instance_unregister(struct device *dev,
 					       struct dual_role_phy_instance
 					       *dual_role){}
-static inline void *dual_role_get_drvdata(struct dual_role_phy_instance
-		*dual_role)
+static void *dual_role_get_drvdata(struct dual_role_phy_instance *dual_role)
 {
 	return ERR_PTR(-ENOSYS);
 }
