diff --git a/include/linux/usb/composite.h b/include/linux/usb/composite.h
index 8fde86d570a1..734146ca9ca5 100644
--- a/include/linux/usb/composite.h
+++ b/include/linux/usb/composite.h
@@ -199,7 +199,11 @@ struct usb_function {
 	 * we can't restructure things to avoid mismatching.
 	 * Related:  unbind() may kfree() but bind() won't...
 	 */
-
+#ifdef CONFIG_USB_ANDROID_SAMSUNG_COMPOSITE
+	int (*set_intf_num)(struct usb_function *f,
+			int intf_num, int index_num);
+	int (*set_config_desc)(int conf_num);
+#endif
 	/* configuration management:  bind/unbind */
 	int			(*bind)(struct usb_configuration *,
 					struct usb_function *);
@@ -208,6 +212,12 @@ struct usb_function {
 	void			(*free_func)(struct usb_function *f);
 	struct module		*mod;
 
+#ifdef CONFIG_USB_CONFIGFS_UEVENT
+	/* Optional function for vendor specific processing */
+	int			(*ctrlrequest)(struct usb_function *,
+					const struct usb_ctrlrequest *);
+#endif
+
 	/* runtime state management */
 	int			(*set_alt)(struct usb_function *,
 					unsigned interface, unsigned alt);
@@ -216,6 +226,8 @@ struct usb_function {
 	void			(*disable)(struct usb_function *);
 	int			(*setup)(struct usb_function *,
 					const struct usb_ctrlrequest *);
+	bool			(*req_match)(struct usb_function *,
+					const struct usb_ctrlrequest *);
 	void			(*suspend)(struct usb_function *);
 	void			(*resume)(struct usb_function *);
 
@@ -228,6 +240,8 @@ struct usb_function {
 	struct list_head		list;
 	DECLARE_BITMAP(endpoints, 32);
 	const struct usb_function_instance *fi;
+
+	unsigned int		bind_deactivated:1;
 };
 
 int usb_add_function(struct usb_configuration *, struct usb_function *);
@@ -489,8 +503,19 @@ struct usb_composite_dev {
 	 */
 	int				delayed_status;
 
+#ifdef CONFIG_USB_ANDROID_SAMSUNG_COMPOSITE
+		/* used by enable_store function of android.c
+		 * to avoid signalling switch changes
+		 */
+	bool				mute_switch;
+	bool				force_disconnect;
+#endif
+
 	/* protects deactivations and delayed_status counts*/
 	spinlock_t			lock;
+
+	unsigned			setup_pending:1;
+	unsigned			os_desc_pending:1;
 };
 
 extern int usb_string_id(struct usb_composite_dev *c);
