diff --git a/include/linux/usb/of.h b/include/linux/usb/of.h
index cfe0528cdbb1..680126b569d0 100644
--- a/include/linux/usb/of.h
+++ b/include/linux/usb/of.h
@@ -15,6 +15,7 @@
 enum usb_dr_mode of_usb_get_dr_mode(struct device_node *np);
 enum usb_device_speed of_usb_get_maximum_speed(struct device_node *np);
 bool of_usb_host_tpl_support(struct device_node *np);
+unsigned int of_usb_get_suspend_clk_freq(struct device_node *np);
 #else
 static inline enum usb_dr_mode of_usb_get_dr_mode(struct device_node *np)
 {
@@ -26,10 +27,16 @@ of_usb_get_maximum_speed(struct device_node *np)
 {
 	return USB_SPEED_UNKNOWN;
 }
+
 static inline bool of_usb_host_tpl_support(struct device_node *np)
 {
 	return false;
 }
+
+static inline unsigned int of_usb_get_suspend_clk_freq(struct device_node *np)
+{
+	return 0;
+}
 #endif
 
 #if IS_ENABLED(CONFIG_OF) && IS_ENABLED(CONFIG_USB_SUPPORT)
