diff --git a/include/linux/usb/otg-fsm.h b/include/linux/usb/otg-fsm.h
index b6ba1bfb86f2..b6e7cbec3371 100644
--- a/include/linux/usb/otg-fsm.h
+++ b/include/linux/usb/otg-fsm.h
@@ -26,8 +26,10 @@
 #ifdef VERBOSE
 #define VDBG(fmt, args...) pr_debug("[%s]  " fmt , \
 				 __func__, ## args)
+/*
 #else
 #define VDBG(stuff...)	do {} while (0)
+*/
 #endif
 
 #ifdef VERBOSE
@@ -59,6 +61,8 @@ enum otg_fsm_timer {
 
 /* OTG state machine according to the OTG spec */
 struct otg_fsm {
+	int reset;
+
 	/* Input */
 	int id;
 	int adp_change;
