diff --git a/drivers/input/evdev.c b/drivers/input/evdev.c
index bd939f252b55..75da2e7b16fc 100644
--- a/drivers/input/evdev.c
+++ b/drivers/input/evdev.c
@@ -190,6 +190,11 @@ static void evdev_pass_values(struct evdev_client *client,
 		__pass_event(client, &event);
 		if (v->type == EV_SYN && v->code == SYN_REPORT)
 			wakeup = true;
+#ifdef CONFIG_USB_HMT_SAMSUNG_INPUT
+		if (v->type== EV_KEY && v->code >= KEY_HMT_CMD_START)
+			pr_info("%s type:KEY code:0x%x value:%x\n", __func__,
+					v->code, v->value);
+#endif
 	}
 
 	spin_unlock(&client->buffer_lock);
