diff --git a/sound/firewire/packets-buffer.c b/sound/firewire/packets-buffer.c
index 3b09b8ef3a09..ea1506679c66 100644
--- a/sound/firewire/packets-buffer.c
+++ b/sound/firewire/packets-buffer.c
@@ -37,7 +37,7 @@ int iso_packets_buffer_init(struct iso_packets_buffer *b, struct fw_unit *unit,
 	packets_per_page = PAGE_SIZE / packet_size;
 	if (WARN_ON(!packets_per_page)) {
 		err = -EINVAL;
-		goto err_packets;
+		goto error;
 	}
 	pages = DIV_ROUND_UP(count, packets_per_page);
 
