diff --git a/net/core/dev.c b/net/core/dev.c
index a45a0c08ea14..65b4687b87cf 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -6333,8 +6333,6 @@ int register_netdevice(struct net_device *dev)
 	ret = notifier_to_errno(ret);
 	if (ret) {
 		rollback_registered(dev);
-		rcu_barrier();
-
 		dev->reg_state = NETREG_UNREGISTERED;
 		/* We should put the kobject that hold in
 		 * netdev_unregister_kobject(), otherwise
@@ -7267,8 +7265,6 @@ static void __net_exit default_device_exit(struct net *net)
 
 		/* Push remaining network devices to init_net */
 		snprintf(fb_name, IFNAMSIZ, "dev%d", dev->ifindex);
-		if (__dev_get_by_name(&init_net, fb_name))
-			snprintf(fb_name, IFNAMSIZ, "dev%%d");
 		err = dev_change_net_namespace(dev, &init_net, fb_name);
 		if (err) {
 			pr_emerg("%s: failed to move %s to init_net: %d\n",
