diff --git a/net/core/fib_rules.c b/net/core/fib_rules.c
index 60fbcdd46fc0..bd5a4bbbfab1 100644
--- a/net/core/fib_rules.c
+++ b/net/core/fib_rules.c
@@ -415,6 +415,7 @@ static int fib_nl_newrule(struct sk_buff *skb, struct nlmsghdr* nlh)
 			unresolved = 1;
 	} else if (rule->action == FR_ACT_GOTO)
 		goto errout_free;
+	
 
 	if (tb[FRA_UID_RANGE]) {
 		if (current_user_ns() != net->user_ns) {
@@ -509,6 +510,7 @@ static int fib_nl_delrule(struct sk_buff *skb, struct nlmsghdr* nlh)
 	err = validate_rulemsg(frh, tb, ops);
 	if (err < 0)
 		goto errout;
+	
 
 	if (tb[FRA_UID_RANGE]) {
 		range = nla_get_kuid_range(tb);
