diff --git a/net/socket.c b/net/socket.c
index 0811517ab6a3..51f8b022547a 100644
--- a/net/socket.c
+++ b/net/socket.c
@@ -2248,11 +2248,10 @@ static int ___sys_recvmsg(struct socket *sock, struct msghdr __user *msg,
 
 	if (MSG_CMSG_COMPAT & flags)
 		err = get_compat_msghdr(msg_sys, msg_compat);
-	else
+	else 
 		err = copy_msghdr_from_user(msg_sys, msg);
-	if (err)
-		return err;
-
+	if (err < 0)
+			return err;
 	if (msg_sys->msg_iovlen > UIO_FASTIOV) {
 		err = -EMSGSIZE;
 		if (msg_sys->msg_iovlen > UIO_MAXIOV)
